--- a/com/yogpc/qp/block/ADismCBlock.java
+++ b/com/yogpc/qp/block/ADismCBlock.java
@@ -11,8 +11,8 @@
 import cofh.api.block.IDismantleable;
 
 public abstract class ADismCBlock extends BlockContainer implements IDismantleable {
-  protected ADismCBlock(final Material m) {
-    super(m);
+  protected ADismCBlock(final int id, final Material m) {
+    super(id, m);
   }
 
   @Override
@@ -22,17 +22,17 @@
   }
 
   @Override
-  public ArrayList<ItemStack> dismantleBlock(final EntityPlayer e, final World w, final int x,
+  public ItemStack dismantleBlock(final EntityPlayer e, final World w, final int x,
       final int y, final int z, final boolean toinv) {
-    final ArrayList<ItemStack> ret = getBlockDropped(w, x, y, z, w.getBlockMetadata(x, y, z), 0);
+    int m = w.getBlockMetadata(x, y, z);
+    final ItemStack ret = new ItemStack(idDropped(m, w.rand, 0), quantityDropped(m, 0, w.rand), damageDropped(m));
     w.setBlockToAir(x, y, z);
-    if (!toinv)
-      for (final ItemStack is : ret) {
+    if (!toinv) {
         final float f = 0.7F;
         final double d0 = w.rand.nextFloat() * f + (1.0F - f) * 0.5D;
         final double d1 = w.rand.nextFloat() * f + (1.0F - f) * 0.5D;
         final double d2 = w.rand.nextFloat() * f + (1.0F - f) * 0.5D;
-        final EntityItem entityitem = new EntityItem(w, x + d0, y + d1, z + d2, is);
+        final EntityItem entityitem = new EntityItem(w, x + d0, y + d1, z + d2, ret);
         entityitem.delayBeforeCanPickup = 10;
         w.spawnEntityInWorld(entityitem);
       }
--- a/com/yogpc/qp/block/BlockBreaker.java
+++ b/com/yogpc/qp/block/BlockBreaker.java
@@ -114,7 +114,7 @@
         w.getBlockMetadata(tx, ty, tz);
     if (ty < 1)
       return;
-    final Block b = w.getBlock(tx, ty, tz);
+    final Block b = Block.blocksList[w.getBlockId(tx, ty, tz)];
     if (b == null || b.isAirBlock(w, x, y, z))
       return;
     final EntityPlayer player = FakePlayerFactory.getMinecraft(w);
@@ -157,22 +157,22 @@
 
   static void setDispenserDefaultDirection(final World w, final int x, final int y, final int z) {
     if (!w.isRemote) {
-      final Block b1 = w.getBlock(x, y, z - 1);
-      final Block b2 = w.getBlock(x, y, z + 1);
-      final Block b3 = w.getBlock(x - 1, y, z);
-      final Block b4 = w.getBlock(x + 1, y, z);
+      final int b1 = w.getBlockId(x, y, z - 1);
+      final int b2 = w.getBlockId(x, y, z + 1);
+      final int b3 = w.getBlockId(x - 1, y, z);
+      final int b4 = w.getBlockId(x + 1, y, z);
       byte b0 = 3;
 
-      if (b1.func_149730_j() && !b2.func_149730_j())
+      if (Block.opaqueCubeLookup[b1] && !Block.opaqueCubeLookup[b2])
         b0 = 3;
 
-      if (b2.func_149730_j() && !b1.func_149730_j())
+      if (Block.opaqueCubeLookup[b2] && !Block.opaqueCubeLookup[b1])
         b0 = 2;
 
-      if (b3.func_149730_j() && !b4.func_149730_j())
+      if (Block.opaqueCubeLookup[b3] && !Block.opaqueCubeLookup[b4])
         b0 = 5;
 
-      if (b4.func_149730_j() && !b3.func_149730_j())
+      if (Block.opaqueCubeLookup[b4] && !Block.opaqueCubeLookup[b3])
         b0 = 4;
       w.setBlockMetadataWithNotify(x, y, z, b0, 2);
     }
@@ -188,7 +188,7 @@
     ItemStack is;
     for (int i = 0; i < count; i++) {
       final int id1 = idDropped(meta, world.rand, 0);
-      if (id1 != null) {
+      if (id1 != 0) {
         is = new ItemStack(id1, 1, damageDropped(meta));
         EnchantmentHelper.enchantmentToIS(tile, is);
         this.drop.add(is);
--- a/com/yogpc/qp/block/BlockController.java
+++ b/com/yogpc/qp/block/BlockController.java
@@ -144,17 +144,17 @@
   }
 
   @Override
-  public ArrayList<ItemStack> dismantleBlock(final EntityPlayer e, final World w, final int x,
+  public ItemStack dismantleBlock(final EntityPlayer e, final World w, final int x,
       final int y, final int z, final boolean toinv) {
-    final ArrayList<ItemStack> ret = getBlockDropped(w, x, y, z, w.getBlockMetadata(x, y, z), 0);
+    int m = w.getBlockMetadata(x, y, z);
+    final ItemStack ret = new ItemStack(idDropped(m, w.rand, 0), quantityDropped(m, 0, w.rand), damageDropped(m));
     w.setBlockToAir(x, y, z);
-    if (!toinv)
-      for (final ItemStack is : ret) {
+    if (!toinv) {
         final float f = 0.7F;
         final double d0 = w.rand.nextFloat() * f + (1.0F - f) * 0.5D;
         final double d1 = w.rand.nextFloat() * f + (1.0F - f) * 0.5D;
         final double d2 = w.rand.nextFloat() * f + (1.0F - f) * 0.5D;
-        final EntityItem entityitem = new EntityItem(w, x + d0, y + d1, z + d2, is);
+        final EntityItem entityitem = new EntityItem(w, x + d0, y + d1, z + d2, ret);
         entityitem.delayBeforeCanPickup = 10;
         w.spawnEntityInWorld(entityitem);
       }
--- a/com/yogpc/qp/block/BlockFrame.java
+++ b/com/yogpc/qp/block/BlockFrame.java
@@ -68,7 +68,7 @@
 
   @Override
   public int idDropped(final int i, final Random random, final int j) {
-    return null;
+    return 0;
   }
 
   @Override
--- a/com/yogpc/qp/block/BlockLaser.java
+++ b/com/yogpc/qp/block/BlockLaser.java
@@ -111,7 +111,7 @@
     ItemStack is;
     for (int i = 0; i < count; i++) {
       final int it = idDropped(meta, world.rand, 0);
-      if (it != null) {
+      if (it != 0) {
         is = new ItemStack(it, 1, damageDropped(meta));
         EnchantmentHelper.enchantmentToIS(tile, is);
         this.drop.add(is);
--- a/com/yogpc/qp/block/BlockMiningWell.java
+++ b/com/yogpc/qp/block/BlockMiningWell.java
@@ -125,7 +125,7 @@
       return;
     final int count = quantityDropped(meta, 0, world.rand);
     final int it = idDropped(meta, world.rand, 0);
-    if (it != null)
+    if (it != 0)
       for (int i = 0; i < count; i++) {
         final ItemStack is = new ItemStack(it, 1, damageDropped(meta));
         EnchantmentHelper.enchantmentToIS(tile, is);
--- a/com/yogpc/qp/block/BlockMover.java
+++ b/com/yogpc/qp/block/BlockMover.java
@@ -77,17 +77,17 @@
   }
 
   @Override
-  public ArrayList<ItemStack> dismantleBlock(final EntityPlayer e, final World w, final int x,
+  public ItemStack dismantleBlock(final EntityPlayer e, final World w, final int x,
       final int y, final int z, final boolean toinv) {
-    final ArrayList<ItemStack> ret = getBlockDropped(w, x, y, z, w.getBlockMetadata(x, y, z), 0);
+    int m = w.getBlockMetadata(x, y, z);
+    final ItemStack ret = new ItemStack(idDropped(m, w.rand, 0), quantityDropped(m, 0, w.rand), damageDropped(m));
     w.setBlockToAir(x, y, z);
-    if (!toinv)
-      for (final ItemStack is : ret) {
+    if (!toinv) {
         final float f = 0.7F;
         final double d0 = w.rand.nextFloat() * f + (1.0F - f) * 0.5D;
         final double d1 = w.rand.nextFloat() * f + (1.0F - f) * 0.5D;
         final double d2 = w.rand.nextFloat() * f + (1.0F - f) * 0.5D;
-        final EntityItem entityitem = new EntityItem(w, x + d0, y + d1, z + d2, is);
+        final EntityItem entityitem = new EntityItem(w, x + d0, y + d1, z + d2, ret);
         entityitem.delayBeforeCanPickup = 10;
         w.spawnEntityInWorld(entityitem);
       }
--- a/com/yogpc/qp/block/BlockPlacer.java
+++ b/com/yogpc/qp/block/BlockPlacer.java
@@ -95,7 +95,7 @@
     final boolean flag1 = (i1 & 8) != 0;
 
     if (flag && !flag1) {
-      par1World.scheduleBlockUpdate(par2, par3, par4, this, 0);
+      par1World.scheduleBlockUpdate(par2, par3, par4, this.blockID, 0);
       par1World.setBlockMetadataWithNotify(par2, par3, par4, i1 | 8, 4);
     } else if (!flag && flag1)
       par1World.setBlockMetadataWithNotify(par2, par3, par4, i1 & -9, 4);
@@ -162,7 +162,7 @@
               tz + fd6.offsetZ, sd5, 0.5F, 0.5F, 0.5F))
             break;
         }
-        final Block k = world.getBlock(tx, ty, tz);
+        final Block k = Block.blocksList[world.getBlockId(tx, ty, tz)];
         if (k != null && k.onBlockActivated(world, tx, ty, tz, player, sd1, 0.5F, 0.5F, 0.5F))
           break;
         if (is != null) {
--- a/com/yogpc/qp/block/BlockPlainPipe.java
+++ b/com/yogpc/qp/block/BlockPlainPipe.java
@@ -30,6 +30,6 @@
 
   @Override
   public int idDropped(final int i, final Random random, final int fortune) {
-    return null;
+    return 0;
   }
 }
--- a/com/yogpc/qp/block/BlockPump.java
+++ b/com/yogpc/qp/block/BlockPump.java
@@ -97,7 +97,7 @@
       return;
     final int count = quantityDropped(meta, 0, world.rand);
     final int it = idDropped(meta, world.rand, 0);
-    if (it != null)
+    if (it != 0)
       for (int i = 0; i < count; i++) {
         final ItemStack is = new ItemStack(it, 1, damageDropped(meta));
         EnchantmentHelper.enchantmentToIS(tile, is);
--- a/com/yogpc/qp/block/BlockQuarry.java
+++ b/com/yogpc/qp/block/BlockQuarry.java
@@ -60,7 +60,7 @@
       return;
     final int count = quantityDropped(meta, 0, world.rand);
     final int it = idDropped(meta, world.rand, 0);
-    if (it != null)
+    if (it != 0)
       for (int i = 0; i < count; i++) {
         final ItemStack is = new ItemStack(it, 1, damageDropped(meta));
         EnchantmentHelper.enchantmentToIS(tile, is);
--- a/com/yogpc/qp/block/BlockRefinery.java
+++ b/com/yogpc/qp/block/BlockRefinery.java
@@ -60,7 +60,7 @@
       return;
     final int count = quantityDropped(meta, 0, world.rand);
     final int it = idDropped(meta, world.rand, 0);
-    if (it != null)
+    if (it != 0)
       for (int i = 0; i < count; i++) {
         final ItemStack is = new ItemStack(it, 1, damageDropped(meta));
         EnchantmentHelper.enchantmentToIS(tile, is);
@@ -87,7 +87,7 @@
   }
 
   private static void consumeItem(final EntityPlayer ep, final ItemStack stack) {
-    final ItemStack container = stack.getItem().getContainerItem();
+    final ItemStack container = stack.getItem().hasContainerItem() ? new ItemStack(stack.getItem().getContainerItem()) : null;
     stack.stackSize--;
     if (stack.stackSize > 0)
       ep.inventory.setInventorySlotContents(ep.inventory.currentItem, stack);
--- a/com/yogpc/qp/ChunkLoadingHandler.java
+++ b/com/yogpc/qp/ChunkLoadingHandler.java
@@ -47,8 +47,8 @@
       final int quarryY = ticket.getModData().getInteger("quarryY");
       final int quarryZ = ticket.getModData().getInteger("quarryZ");
 
-      final Block b = world.getBlock(quarryX, quarryY, quarryZ);
-      if (b == QuarryPlusI.blockQuarry || b == QuarryPlusI.blockMarker)
+      final int b = world.getBlockId(quarryX, quarryY, quarryZ);
+      if (b == QuarryPlusI.blockQuarry.blockID || b == QuarryPlusI.blockMarker.blockID)
         validTickets.add(ticket);
     }
     return validTickets;
--- a/com/yogpc/qp/compat/BuildCraftHelper.java
+++ b/com/yogpc/qp/compat/BuildCraftHelper.java
@@ -7,7 +7,7 @@
 public class BuildCraftHelper {
   public static final boolean isWrench(final Item equipped, final EntityPlayer ep, final int x,
       final int y, final int z) {
-    if (ModAPIManager.INSTANCE.hasAPI("BuildCraftAPI|tools") && equipped instanceof IToolWrench
+    if (equipped instanceof IToolWrench
         && ((IToolWrench) equipped).canWrench(ep, x, y, z)) {
       ((IToolWrench) equipped).wrenchUsed(ep, x, y, z);
       return true;
--- a/com/yogpc/qp/compat/RefineryRecipeHelper.java
+++ b/com/yogpc/qp/compat/RefineryRecipeHelper.java
@@ -39,12 +39,9 @@
   }
 
   public static void get(final TileRefinery tr) {
-    if (!ModAPIManager.INSTANCE.hasAPI("BuildCraftAPI|recipes") || tr.cached != null)
-      return;
-    final RefineryRecipes.RecipeManager irrm = BuildcraftRecipes.refinery;
-    if (irrm == null)
+    if (tr.cached != null)
       return;
-    for (final RefineryRecipes.Recipe ifr : irrm.getRecipes()) {
+    for (final RefineryRecipes.Recipe ifr : RefineryRecipes.getRecipes()) {
       if (!craft(ifr, tr, false) || !RefineryRecipeHelper.check(ifr.result, tr))
         continue;
       craft(ifr, tr, true);
--- a/com/yogpc/qp/container/ContainerMover.java
+++ b/com/yogpc/qp/container/ContainerMover.java
@@ -74,7 +74,7 @@
 
   @Override
   public boolean canInteractWith(final EntityPlayer var1) {
-    return this.worldObj.getBlock(this.posX, this.posY, this.posZ) != QuarryPlusI.blockMover ? false
+    return this.worldObj.getBlockId(this.posX, this.posY, this.posZ) != QuarryPlusI.blockMover.blockID ? false
         : var1.getDistanceSq(this.posX + 0.5D, this.posY + 0.5D, this.posZ + 0.5D) <= 64.0D;
   }
 
@@ -164,7 +164,7 @@
             final NBTTagList nlist = new NBTTagList();
             for (int j = 0; j < list.tagCount(); j++)
               if (((NBTTagCompound)list.tagAt(j)).getShort("id") != this.avail)
-                nlist.appendTag(((NBTTagCompound)list.tagAt(j)));
+                nlist.appendTag((list.tagAt(j)));
             list = nlist;
           }
           is.getTagCompound().removeTag("ench");
--- a/com/yogpc/qp/container/ContainerWorkbench.java
+++ b/com/yogpc/qp/container/ContainerWorkbench.java
@@ -31,7 +31,7 @@
     final short id = bb.readShort();
     if (id < 0)
       return null;
-    final ItemStack is = new ItemStack(Item.getItemById(id), bb.readInt(), bb.readShort());
+    final ItemStack is = new ItemStack(id, bb.readInt(), bb.readShort());
     final short nbtlen = bb.readShort();
     if (nbtlen >= 0) {
       final byte[] ba = new byte[nbtlen];
@@ -47,7 +47,7 @@
       bb.writeShort((short) -1);
       return;
     }
-    bb.writeShort((short) Item.getIdFromItem(is.getItem()));
+    bb.writeShort((short) is.itemID);
     bb.writeInt(is.stackSize);
     bb.writeShort((short) is.getItemDamage());
     NBTTagCompound nbttagcompound = null;
--- a/com/yogpc/qp/CreativeTabQuarryPlus.java
+++ b/com/yogpc/qp/CreativeTabQuarryPlus.java
@@ -11,7 +11,7 @@
 
   @Override
   public Item getTabIconItem() {
-    return Item.getItemFromBlock(QuarryPlusI.blockQuarry);
+    return Item.itemsList[QuarryPlusI.blockQuarry.blockID];
   }
 
 }
--- a/com/yogpc/qp/InvUtils.java
+++ b/com/yogpc/qp/InvUtils.java
@@ -99,7 +99,7 @@
         pi.add((IInventory) t);
         pid.add(d.getOpposite());
       }
-      if (ModAPIManager.INSTANCE.hasAPI("BuildCraftAPI|transport") && t instanceof IPipeTile) {
+      if (t instanceof IPipeTile) {
         final IPipeTile p = (IPipeTile) t;
         if (p.getPipeType() != IPipeTile.PipeType.ITEM || !p.isPipeConnected(d.getOpposite()))
           continue;
--- a/com/yogpc/qp/item/ItemArmorElectric.java
+++ b/com/yogpc/qp/item/ItemArmorElectric.java
@@ -47,9 +47,6 @@
     return "yogpstop_qp:textures/models/armor/elecArmor_layer_1.png";
   }
 
-  private static final Field f = ReflectionHelper.getField(NetHandlerPlayServer.class,
-      "field_147365_f", "floatingTickCount");
-
   private static final double maxFallSpeedOnHover = 0.35;
 
   private void useJetpack(final EntityPlayer p, final ItemStack jetpack, final boolean hover) {
@@ -87,11 +84,7 @@
         p.fallDistance = 0.0f;
         p.distanceWalkedModified = 0.0f;
         if (p instanceof EntityPlayerMP)
-          try {
-            f.setInt(((EntityPlayerMP) p).playerNetServerHandler, 0);
-          } catch (final Exception e) {
-            throw new RuntimeException(e);
-          }
+          ((EntityPlayerMP) p).playerNetServerHandler.ticksForFloatKick = 0;
       }
       p.inventoryContainer.detectAndSendChanges();
     }
--- a/com/yogpc/qp/item/ItemTool.java
+++ b/com/yogpc/qp/item/ItemTool.java
@@ -79,7 +79,7 @@
             f = true;
         }
       NBTTagCompound c = is.getTagCompound();
-      final Block b = w.getBlock(x, y, z);
+      final Block b = Block.blocksList[w.getBlockId(x, y, z)];
       BlockData bd = null;
       if (c != null && c.hasKey("Bname")) {
         bd = new BlockData(c.getString("Bname"), c.getInteger("Bmeta"));
--- a/com/yogpc/qp/PacketHandler.java
+++ b/com/yogpc/qp/PacketHandler.java
@@ -41,8 +41,7 @@
 import com.yogpc.qp.container.IPacketContainer;
 import com.yogpc.qp.tile.APacketTile;
 
-@Sharable
-public class PacketHandler extends SimpleChannelInboundHandler<YogpstopPacket> {
+public class PacketHandler implements IPacketHandler {
   @Retention(RetentionPolicy.RUNTIME)
   @Target(ElementType.METHOD)
   public static @interface Handler {
@@ -56,7 +55,6 @@
       registeredStaticHandlers.put(c.getName(), l.get(0));
   }
 
-  public static EnumMap<Side, FMLEmbeddedChannel> channels;
   static final byte Tile = 0;
   static final byte NBT = 1;
   static final byte WINPROP = 2;
@@ -88,14 +86,19 @@
   public static final byte remove_laser = 1;
 
   @Override
-  protected void channelRead0(final ChannelHandlerContext ctx, final YogpstopPacket packet)
-      throws Exception {
+  public void onPacketData(final INetworkManager network, final Packet250CustomPayload _packet, final Player player) {
+    final YogpstopPacket packet = new YogpstopPacket();
+    packet.readData(_packet.data, player);
     if (packet.getChannel() == NBT)
       setNBTFromPacket(packet);
     else if (packet.getChannel() == WINPROP) {
       final Container container = packet.getPlayer().openContainer;
       if (container instanceof IPacketContainer && container.windowId == packet.getHeader()[0])
-        ((IPacketContainer) container).receivePacket(packet.getData());
+        try {
+          ((IPacketContainer) container).receivePacket(packet.getData());
+        } catch (final IOException e) {
+          throw new RuntimeException(e);
+        }
     } else if (packet.getChannel() == Tile) {
       final ByteArrayDataInput hdr = ByteStreams.newDataInput(packet.getHeader());
       final TileEntity t =
@@ -115,33 +118,29 @@
           | packet.getData()[1] << 16 | packet.getData()[2] << 8 | packet.getData()[3]);
   }
 
+  public static Packet250CustomPayload getPacket(final YogpstopPacket p) {
+    final Packet250CustomPayload pkt = new Packet250CustomPayload();
+    pkt.channel = "QuarryPlus";
+    pkt.data = p.writeData();
+    pkt.length = pkt.data.length;
+    return pkt;
+  }
+
   public static void sendPacketToServer(final YogpstopPacket p) {
-    channels.get(Side.CLIENT).attr(FMLOutboundHandler.FML_MESSAGETARGET)
-        .set(OutboundTarget.TOSERVER);
-    channels.get(Side.CLIENT).writeOutbound(p);
+    PacketDispatcher.sendPacketToServer(getPacket(p));
   }
 
   public static void sendPacketToAround(final YogpstopPacket p, final int d, final int x,
       final int y, final int z) {
-    channels.get(Side.SERVER).attr(FMLOutboundHandler.FML_MESSAGETARGET)
-        .set(OutboundTarget.ALLAROUNDPOINT);
-    channels.get(Side.SERVER).attr(FMLOutboundHandler.FML_MESSAGETARGETARGS)
-        .set(new NetworkRegistry.TargetPoint(d, x, y, z, 256));
-    channels.get(Side.SERVER).writeOutbound(p);
+    PacketDispatcher.sendPacketToAllAround(x, y, z, 256, d, getPacket(p));
   }
 
   public static void sendPacketToDimension(final YogpstopPacket p, final int d) {
-    channels.get(Side.SERVER).attr(FMLOutboundHandler.FML_MESSAGETARGET)
-        .set(OutboundTarget.DIMENSION);
-    channels.get(Side.SERVER).attr(FMLOutboundHandler.FML_MESSAGETARGETARGS).set(new Integer(d));
-    channels.get(Side.SERVER).writeOutbound(p);
+    PacketDispatcher.sendPacketToAllInDimension(getPacket(p), d);
   }
 
   public static void sendPacketToPlayer(final YogpstopPacket p, final EntityPlayer e) {
-    channels.get(Side.SERVER).attr(FMLOutboundHandler.FML_MESSAGETARGET)
-        .set(FMLOutboundHandler.OutboundTarget.PLAYER);
-    channels.get(Side.SERVER).attr(FMLOutboundHandler.FML_MESSAGETARGETARGS).set(e);
-    channels.get(Side.SERVER).writeOutbound(p);
+    PacketDispatcher.sendPacketToPlayer(getPacket(p), (Player) e);
   }
 
   private static void setNBTFromPacket(final YogpstopPacket p) {
--- a/com/yogpc/qp/ProxyClient.java
+++ b/com/yogpc/qp/ProxyClient.java
@@ -33,25 +33,44 @@
 import cpw.mods.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class ProxyClient extends ProxyCommon {
+public class ProxyClient extends ProxyCommon implements ITickHandler {
   private int key = 0;
 
   static {
+    final List<KeyBinding> bs = new ArrayList<KeyBinding>();
     for (final Key k : Key.values())
       if (k.name != null) {
-        k.binding = new KeyBinding(k.name, k.id, "key.yoglib");
-        ClientRegistry.registerKeyBinding((KeyBinding) k.binding);
+        k.binding = new KeyBinding(k.name, k.id);
+        bs.add((KeyBinding) k.binding);
       }
+    final GameSettings set = Minecraft.getMinecraft().gameSettings;
+    final KeyBinding[] add = bs.toArray(new KeyBinding[bs.size()]);
+    final KeyBinding[] all = new KeyBinding[set.keyBindings.length + add.length];
+    System.arraycopy(set.keyBindings, 0, all, 0, set.keyBindings.length);
+    System.arraycopy(add, 0, all, set.keyBindings.length, add.length);
+    set.keyBindings = all;
+    set.loadOptions();
   }
 
   public ProxyClient() {
-    FMLCommonHandler.instance().bus().register(this);
+    TickRegistry.registerTickHandler(this, Side.CLIENT);
   }
 
-  @ForgeSubscribe
-  public void keyUpdate(final TickEvent.ClientTickEvent e) {
-    if (e.phase != TickEvent.Phase.START)
-      return;
+  @Override
+  public String getLabel() {
+    return "QuarryPlus";
+  }
+
+  @Override
+  public EnumSet<TickType> ticks() {
+    return EnumSet.of(TickType.CLIENT);
+  }
+
+  @Override
+  public void tickEnd(final EnumSet<TickType> type, final Object... data) {}
+
+  @Override
+  public void tickStart(final EnumSet<TickType> type, final Object... data) {
     final int prev = this.key;
     this.key = 0;
     final GuiScreen currentScreen = Minecraft.getMinecraft().currentScreen;
@@ -80,13 +99,6 @@
   }
 
   @Override
-  public EntityPlayer getPacketPlayer(final INetHandler inh) {
-    if (inh instanceof NetHandlerPlayServer)
-      return ((NetHandlerPlayServer) inh).playerEntity;
-    return Minecraft.getMinecraft().thePlayer;
-  }
-
-  @Override
   public int addNewArmourRendererPrefix(final String s) {
     return RenderingRegistry.addNewArmourRendererPrefix(s);
   }
--- a/com/yogpc/qp/ProxyCommon.java
+++ b/com/yogpc/qp/ProxyCommon.java
@@ -41,12 +41,6 @@
     }
   }
 
-  public EntityPlayer getPacketPlayer(final INetHandler inh) {
-    if (inh instanceof NetHandlerPlayServer)
-      return ((NetHandlerPlayServer) inh).playerEntity;
-    return null;
-  }
-
   public int addNewArmourRendererPrefix(final String s) {
     return 0;
   }
--- a/com/yogpc/qp/QuarryPlus.java
+++ b/com/yogpc/qp/QuarryPlus.java
@@ -24,6 +24,8 @@
     name = "QuarryPlus",
     version = "{version}",
     dependencies = "after:BuildCraftAPI|core;after:BuildCraftAPI|power;after:BuildCraftAPI|recipes;after:BuildCraftAPI|tools;after:BuildCraftAPI|transport;after:CoFHAPI|block;after:CoFHAPI|energy;after:CoFHAPI|inventory;after:IC2API")
+@NetworkMod(clientSideRequired = true, serverSideRequired = false, channels = {"QuarryPlus"},
+    packetHandler = PacketHandler.class)
 public class QuarryPlus {
   @SidedProxy(clientSide = "com.yogpc.qp.ProxyClient", serverSide = "com.yogpc.qp.ProxyCommon")
   public static ProxyCommon proxy;
--- a/com/yogpc/qp/QuarryPlusI.java
+++ b/com/yogpc/qp/QuarryPlusI.java
@@ -134,10 +134,92 @@
     }
   }
 
+  private static final Pattern lang = Pattern.compile("assets/(.*)/lang/(.*)\\.lang");
+  private static final Pattern format = Pattern.compile("%(\\d+\\$)?[\\d\\.]*[df]");
+
+  public static Properties parseLangFile(final InputStream is, final boolean cl) {
+    final Properties table = new Properties();
+    try {
+      final Reader r = new InputStreamReader(is, Charsets.UTF_8);
+      final BufferedReader br = new BufferedReader(r);
+      String s;
+      while ((s = br.readLine()) != null)
+        if (!s.isEmpty() && s.charAt(0) != '#') {
+          final String[] astring = s.split("=", 2);
+          if (astring.length == 2)
+            table.put(astring[0], format.matcher(astring[1]).replaceAll("%$1s"));
+        }
+      if (cl) {
+        br.close();
+        r.close();
+      }
+    } catch (final IOException e) {
+      throw new RuntimeException(e);
+    }
+    return table;
+  }
+
+  private static void loadLangs_dir(final File b, final File f) {
+    String n = f.getPath().replace(b.getPath(), "").replace(File.separator, "/");
+    if (n.length() > 0 && n.charAt(0) == '/')
+      n = n.substring(1);
+    final Matcher m = lang.matcher(n);
+    if (m.matches())
+      try {
+        final InputStream is = new FileInputStream(f);
+        LanguageRegistry.instance().addStringLocalization(parseLangFile(is, true), m.group(2));
+        is.close();
+      } catch (final IOException e) {
+        throw new RuntimeException(e);
+      }
+    if (f.isDirectory())
+      for (final File e : f.listFiles())
+        loadLangs_dir(b, e);
+  }
+
+  public static void loadLangs(final File in) {
+    if (in.isDirectory()) {
+      loadLangs_dir(in, in);
+      return;
+    }
+    try {
+      final InputStream is = new FileInputStream(in);
+      final ZipInputStream zis = new ZipInputStream(is);
+      ZipEntry entry;
+      while ((entry = zis.getNextEntry()) != null) {
+        final Matcher m = lang.matcher(entry.getName());
+        if (m.matches())
+          LanguageRegistry.instance().addStringLocalization(parseLangFile(zis, false), m.group(2));
+        zis.closeEntry();
+      }
+      zis.close();
+      is.close();
+    } catch (final IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
   public static void preInit(final FMLPreInitializationEvent event) {
     final Configuration cfg = new Configuration(event.getSuggestedConfigurationFile());
+    final int[] id = new int[16];
     try {
       cfg.load();
+      id[0] = cfg.getBlock("Workbench", 3333).getInt();
+      id[1] = cfg.getBlock("Controller", 3334).getInt();
+      id[2] = cfg.getItem("MagicMirror", 21477).getInt();
+      id[3] = cfg.getItem("ElectricArmor", 21478).getInt();
+      id[4] = cfg.getBlock("Quarry", 1970).getInt();
+      id[5] = cfg.getBlock("Marker", 1971).getInt();
+      id[6] = cfg.getBlock("EnchantMover", 1972).getInt();
+      id[7] = cfg.getBlock("MiningWell", 1973).getInt();
+      id[8] = cfg.getBlock("Pump", 1974).getInt();
+      id[9] = cfg.getBlock("Refinery", 1976).getInt();
+      id[10] = cfg.getBlock("Placer", 1977).getInt();
+      id[11] = cfg.getBlock("Breaker", 1978).getInt();
+      id[12] = cfg.getBlock("Laser", 1979).getInt();
+      id[13] = cfg.getBlock("PlainPipe", 1980).getInt();
+      id[14] = cfg.getBlock("Frame", 1981).getInt();
+      id[15] = cfg.getItem("Tools", 18463).getInt();
       disableController =
           cfg.get(Configuration.CATEGORY_GENERAL, "DisableSpawnerController", false).getBoolean(
               false);
@@ -151,31 +233,29 @@
     } finally {
       cfg.save();
     }
-    workbench = new BlockWorkbench();
+    workbench = new BlockWorkbench(id[0]);
     if (!disableController)
-      controller = new BlockController();
-    blockQuarry = new BlockQuarry();
-    blockMarker = new BlockMarker();
-    blockMover = new BlockMover();
-    blockMiningWell = new BlockMiningWell();
-    blockPump = new BlockPump();
-    blockRefinery = new BlockRefinery();
-    blockPlacer = new BlockPlacer();
-    blockBreaker = new BlockBreaker();
-    blockLaser = new BlockLaser();
-    blockPlainPipe = new BlockPlainPipe();
-    blockFrame = new BlockFrame();
-    magicmirror = new ItemMirror();
-    armor = new ItemArmorElectric();
-    itemTool = new ItemTool();
+      controller = new BlockController(id[1]);
+    blockQuarry = new BlockQuarry(id[4]);
+    blockMarker = new BlockMarker(id[5]);
+    blockMover = new BlockMover(id[6]);
+    blockMiningWell = new BlockMiningWell(id[7]);
+    blockPump = new BlockPump(id[8]);
+    blockRefinery = new BlockRefinery(id[9]);
+    blockPlacer = new BlockPlacer(id[10]);
+    blockBreaker = new BlockBreaker(id[11]);
+    blockLaser = new BlockLaser(id[12]);
+    blockPlainPipe = new BlockPlainPipe(id[13]);
+    blockFrame = new BlockFrame(id[14]);
+    magicmirror = new ItemMirror(id[2]);
+    armor = new ItemArmorElectric(id[3]);
+    itemTool = new ItemTool(id[15]);
     ForgeChunkManager.setForcedChunkLoadingCallback(QuarryPlus.I, new ChunkLoadingHandler());
     MinecraftForge.EVENT_BUS.register(I);
+    loadLangs(event.getSourceFile());
   }
 
   public static void init() {
-    PacketHandler.channels =
-        NetworkRegistry.instance().newChannel("QuarryPlus", new YogpstopPacketCodec(),
-            new PacketHandler());
     final WeightedRandomChestContent c =
         new WeightedRandomChestContent(new ItemStack(magicmirror), 1, 1, 9);
     ChestGenHooks.getInfo(ChestGenHooks.DUNGEON_CHEST).addItem(c);
--- a/com/yogpc/qp/render/RenderRefinery.java
+++ b/com/yogpc/qp/render/RenderRefinery.java
@@ -203,8 +203,8 @@
     diplayLists = new int[100];
     Block baseBlock;
     Icon texture;
-    if (fluid.getBlock() != null) {
-      baseBlock = fluid.getBlock();
+    if (fluid.getBlockID() > -1) {
+      baseBlock = Block.blocksList[fluid.getBlockID()];
       texture = fluid.getStillIcon();
     } else {
       baseBlock = Block.waterStill;
--- a/com/yogpc/qp/tile/APacketTile.java
+++ b/com/yogpc/qp/tile/APacketTile.java
@@ -26,9 +26,6 @@
 
   @Override
   public final Packet getDescriptionPacket() {
-    final ByteBuf buf = Unpooled.buffer();
-    buf.writeByte(0);
-    new YogpstopPacket(this).writeData(buf);
-    return new FMLProxyPacket(buf, "QuarryPlus");
+    return PacketHandler.getPacket(new YogpstopPacket(this));
   }
 }
--- a/com/yogpc/qp/tile/APowerTile.java
+++ b/com/yogpc/qp/tile/APowerTile.java
@@ -23,15 +26,12 @@
   @Override
   public void updateEntity() {
     super.updateEntity();
-    if (ModAPIManager.INSTANCE.hasAPI("IC2API") && !this.ic2ok && !this.worldObj.isRemote) {
-      try {
-        IC2Helper.ic2load(this);
-      } catch (final Exception e) {
-      }
+    if (!this.ic2ok && !this.worldObj.isRemote) {
+      MinecraftForge.EVENT_BUS.post(new EnergyTileLoadEvent(this));
       this.ic2ok = true;
     }
     final double rem = Math.min(this.maxGot - this.got, this.max - this.all - this.got);
-    this.got += this.pp.useEnergy(0, rem, true);
+    this.got += this.pp.useEnergy(0, (float) rem, true);
     this.all += this.got;
     this.got = 0;
   }
@@ -46,10 +46,7 @@
   public void onChunkUnload() {
     super.onChunkUnload();
     if (this.ic2ok && !this.worldObj.isRemote) {
-      try {
-        IC2Helper.ic2unload(this);
-      } catch (final Exception e) {
-      }
+      MinecraftForge.EVENT_BUS.post(new EnergyTileUnloadEvent(this));
       this.ic2ok = false;
     }
   }
@@ -60,7 +57,7 @@
     this.all = nbttc.getDouble("storedEnergy");
     this.max = nbttc.getDouble("MAX_stored");
     this.maxGot = nbttc.getDouble("MAX_receive");
-    this.pp.configure(0, this.maxGot, 0, this.max);
+    this.pp.configure(0, (float) this.maxGot, 0, (float) this.max);
   }
 
   @Override
@@ -105,7 +102,7 @@
   public final void configure(final double x, final double maxstored) {
     this.maxGot = x;
     this.max = maxstored;
-    this.pp.configure(0, this.maxGot, 0, this.max);
+    this.pp.configure(0, (float) this.maxGot, 0, (float) this.max);
   }
 
   @Override
--- a/com/yogpc/qp/tile/TileBasic.java
+++ b/com/yogpc/qp/tile/TileBasic.java
@@ -39,8 +39,7 @@
 import com.yogpc.qp.QuarryPlusI.BlockData;
 import com.yogpc.qp.ReflectionHelper;
 
-public abstract class TileBasic extends APowerTile implements IEnchantableTile, IInventory,
-    IInventoryConnection {
+public abstract class TileBasic extends APowerTile implements IEnchantableTile, IInventory {
   protected ForgeDirection pump = ForgeDirection.UNKNOWN;
 
   public final List<BlockData> fortuneList = new ArrayList<BlockData>();
@@ -99,8 +98,8 @@
 
   protected boolean S_breakBlock(final int x, final int y, final int z) {
     final Collection<ItemStack> dropped = new LinkedList<ItemStack>();
-    final Block b =
-        this.worldObj.getChunkProvider().loadChunk(x >> 4, z >> 4).getBlock(x & 0xF, y, z & 0xF);
+    final Block b =Block.blocksList[
+        this.worldObj.getChunkProvider().loadChunk(x >> 4, z >> 4).getBlockID(x & 0xF, y, z & 0xF)];
     if (b == null || b.isAirBlock(this.worldObj, x, y, z))
       return true;
     if (TilePump.isLiquid(b, false, null, 0, 0, 0, 0)) {
@@ -140,8 +139,7 @@
     final int meta = this.worldObj.getBlockMetadata(x, y, z);
     if (b.canSilkHarvest(this.worldObj, null, x, y, z, meta)
         && this.silktouch
-        && this.silktouchList.contains(new BlockData(GameData.getBlockRegistry()
-            .getNameForObject(b), meta)) == this.silktouchInclude) {
+        && this.silktouchList.contains(new BlockData(Integer.toString(b.blockID), meta)) == this.silktouchInclude) {
       list.add((ItemStack) ReflectionHelper.invoke(createStackedBlock, b, new Integer(meta)));
       return -1;
     }
@@ -301,9 +299,4 @@
   public boolean isItemValidForSlot(final int p_94041_1_, final ItemStack p_94041_2_) {
     return false;
   }
-
-  @Override
-    public ConnectionType canConnectInventory(final ForgeDirection arg0) {
-    return ConnectionType.FORCE;
-  }
 }
--- a/com/yogpc/qp/tile/TileBreaker.java
+++ b/com/yogpc/qp/tile/TileBreaker.java
@@ -25,8 +25,7 @@
 import net.minecraft.tileentity.TileEntity;
 import net.minecraftforge.common.ForgeDirection;
 
-public class TileBreaker extends TileEntity implements IEnchantableTile, IInventory,
-    IInventoryConnection {
+public class TileBreaker extends TileEntity implements IEnchantableTile, IInventory {
   public boolean silktouch = false;
   public byte fortune = 0;
 
@@ -121,9 +120,4 @@
   public boolean isItemValidForSlot(final int p_94041_1_, final ItemStack p_94041_2_) {
     return false;
   }
-
-  @Override
-    public ConnectionType canConnectInventory(final ForgeDirection arg0) {
-    return ConnectionType.FORCE;
-  }
 }
--- a/com/yogpc/qp/tile/TileMarker.java
+++ b/com/yogpc/qp/tile/TileMarker.java
@@ -199,7 +199,7 @@
         final boolean bb) {
       final ArrayList<ItemStack> ret = new ArrayList<ItemStack>();
       final TileEntity te = this.w.getBlockTileEntity(x, y, z);
-      final Block b = this.w.getBlock(x, y, z);
+      final Block b = Block.blocksList[this.w.getBlockId(x, y, z)];
       if (b instanceof BlockMarker) {
         if (te instanceof TileMarker)
           ((TileMarker) te).link = null;
@@ -415,16 +415,16 @@
   private static void S_renewConnection(final Link l, final World w, final int x, final int y,
       final int z) {
     int tx = 0, ty = 0, tz = 0;
-    Block b;
+    int b;
     if (l.xx == l.xn) {
       for (tx = 1; tx <= MAX_SIZE; tx++) {
-        b = w.getBlock(x + tx, y, z);
-        if (b instanceof BlockMarker && !linkList.contains(new BlockIndex(w, x + tx, y, z))) {
+        b = w.getBlockId(x + tx, y, z);
+        if (b == QuarryPlusI.blockMarker.blockID && !linkList.contains(new BlockIndex(w, x + tx, y, z))) {
           l.xx = x + tx;
           break;
         }
-        b = w.getBlock(x - tx, y, z);
-        if (b instanceof BlockMarker && !linkList.contains(new BlockIndex(w, x - tx, y, z))) {
+        b = w.getBlockId(x - tx, y, z);
+        if (b == QuarryPlusI.blockMarker.blockID && !linkList.contains(new BlockIndex(w, x - tx, y, z))) {
           tx = -tx;
           l.xn = x + tx;
           break;
@@ -435,13 +435,13 @@
     }
     if (l.yx == l.yn) {
       for (ty = 1; ty <= MAX_SIZE; ty++) {
-        b = w.getBlock(x, y + ty, z);
-        if (b instanceof BlockMarker && !linkList.contains(new BlockIndex(w, x, y + ty, z))) {
+        b = w.getBlockId(x, y + ty, z);
+        if (b == QuarryPlusI.blockMarker.blockID && !linkList.contains(new BlockIndex(w, x, y + ty, z))) {
           l.yx = y + ty;
           break;
         }
-        b = w.getBlock(x, y - ty, z);
-        if (b instanceof BlockMarker && !linkList.contains(new BlockIndex(w, x, y - ty, z))) {
+        b = w.getBlockId(x, y - ty, z);
+        if (b == QuarryPlusI.blockMarker.blockID && !linkList.contains(new BlockIndex(w, x, y - ty, z))) {
           ty = -ty;
           l.yn = y + ty;
           break;
@@ -452,13 +452,13 @@
     }
     if (l.zx == l.zn) {
       for (tz = 1; tz <= MAX_SIZE; tz++) {
-        b = w.getBlock(x, y, z + tz);
-        if (b instanceof BlockMarker && !linkList.contains(new BlockIndex(w, x, y, z + tz))) {
+        b = w.getBlockId(x, y, z + tz);
+        if (b == QuarryPlusI.blockMarker.blockID && !linkList.contains(new BlockIndex(w, x, y, z + tz))) {
           l.zx = z + tz;
           break;
         }
-        b = w.getBlock(x, y, z - tz);
-        if (b instanceof BlockMarker && !linkList.contains(new BlockIndex(w, x, y, z - tz))) {
+        b = w.getBlockId(x, y, z - tz);
+        if (b == QuarryPlusI.blockMarker.blockID && !linkList.contains(new BlockIndex(w, x, y, z - tz))) {
           tz = -tz;
           l.zn = z + tz;
           break;
@@ -628,7 +628,7 @@
   @Override
   public void invalidate() {
     super.invalidate();
-    if (this.worldObj.getBlock(this.xCoord, this.yCoord, this.zCoord) != QuarryPlusI.blockMarker)
+    if (this.worldObj.getBlockId(this.xCoord, this.yCoord, this.zCoord) != QuarryPlusI.blockMarker.blockID)
       G_destroy();
   }
 }
--- a/com/yogpc/qp/tile/TileMiningWell.java
+++ b/com/yogpc/qp/tile/TileMiningWell.java
@@ -65,7 +65,7 @@
     int depth = this.yCoord - 1;
     while (!S_checkTarget(depth)) {
       if (this.working)
-        this.worldObj.setBlock(this.xCoord, depth, this.zCoord, QuarryPlusI.blockPlainPipe);
+        this.worldObj.setBlock(this.xCoord, depth, this.zCoord, QuarryPlusI.blockPlainPipe.blockID);
       depth--;
     }
     if (this.working)
@@ -78,15 +78,15 @@
       G_destroy();
       return true;
     }
-    final Block b =
+    final Block b = Block.blocksList[
         this.worldObj.getChunkProvider().loadChunk(this.xCoord >> 4, this.zCoord >> 4)
-            .getBlock(this.xCoord & 0xF, depth, this.zCoord & 0xF);
+            .getBlockID(this.xCoord & 0xF, depth, this.zCoord & 0xF)];
     final float h =
         b == null ? -1 : b.getBlockHardness(this.worldObj, this.xCoord, depth, this.zCoord);
     if (b == null || h < 0 || b == QuarryPlusI.blockPlainPipe
         || b.isAirBlock(this.worldObj, this.xCoord, depth, this.zCoord))
       return false;
-    if (this.pump == ForgeDirection.UNKNOWN && b.getMaterial().isLiquid())
+    if (this.pump == ForgeDirection.UNKNOWN && this.worldObj.getBlockMaterial(this.xCoord, depth, this.zCoord).isLiquid())
       return false;
     if (!this.working) {
       this.working = true;
@@ -125,7 +125,7 @@
     G_renew_powerConfigure();
     PacketHandler.sendNowPacket(this, (byte) 0);
     for (int depth = this.yCoord - 1; depth > 0; depth--) {
-      if (this.worldObj.getBlock(this.xCoord, depth, this.zCoord) != QuarryPlusI.blockPlainPipe)
+      if (this.worldObj.getBlockId(this.xCoord, depth, this.zCoord) != QuarryPlusI.blockPlainPipe.blockID)
         break;
       this.worldObj.setBlockToAir(this.xCoord, depth, this.zCoord);
     }
--- a/com/yogpc/qp/tile/TilePlacer.java
+++ b/com/yogpc/qp/tile/TilePlacer.java
@@ -21,7 +21,7 @@
 import net.minecraft.tileentity.TileEntity;
 import net.minecraftforge.common.ForgeDirection;
 
-public class TilePlacer extends TileEntity implements IInventory, IInventoryConnection {
+public class TilePlacer extends TileEntity implements IInventory {
   private final ItemStack[] stack = new ItemStack[getSizeInventory()];
 
   @Override
@@ -128,9 +128,4 @@
       }
     par1NBTTagCompound.setTag("Items", nbttaglist);
   }
-
-  @Override
-    public ConnectionType canConnectInventory(final ForgeDirection arg0) {
-    return ConnectionType.FORCE;
-  }
 }
--- a/com/yogpc/qp/tile/TilePump.java
+++ b/com/yogpc/qp/tile/TilePump.java
@@ -84,7 +84,7 @@
     if (nbttc.getTag("mapping0") instanceof NBTTagList)
       for (int i = 0; i < this.mapping.length; i++)
         readStringCollection(
-            nbttc.getTagList(new StringBuilder().append("mapping").append(i).toString(), 8),
+            nbttc.getTagList(new StringBuilder().append("mapping").append(i).toString()),
             this.mapping[i]);
     this.range = nbttc.getByte("range");
     this.quarryRange = nbttc.getBoolean("quarryRange");
@@ -409,7 +409,7 @@
     while (cp != cg) {
       ebs_c = this.ebses[xb[cg] >> 4][zb[cg] >> 4][yb[cg] >> 4];
       if (ebs_c != null) {
-        b_c = ebs_c.getBlockByExtId(xb[cg] & 0xF, yb[cg] & 0xF, zb[cg] & 0xF);
+        b_c = Block.blocksList[ebs_c.getExtBlockID(xb[cg] & 0xF, yb[cg] & 0xF, zb[cg] & 0xF)];
         if (this.blocks[yb[cg] - this.yOffset][xb[cg]][zb[cg]] == 0
             && isLiquid(b_c, false, null, 0, 0, 0, 0)) {
           this.blocks[yb[cg] - this.yOffset][xb[cg]][zb[cg]] = 0x3F;
@@ -470,8 +470,7 @@
             for (bz = 0; bz < this.block_side_z; bz++)
               if ((this.blocks[this.py - this.yOffset][bx][bz] & 0x40) != 0) {
                 bb =
-                    this.ebses[bx >> 4][bz >> 4][this.py >> 4].getBlockByExtId(bx & 0xF,
-                        this.py & 0xF, bz & 0xF);
+                    Block.blocksList[this.ebses[bx >> 4][bz >> 4][this.py >> 4].getExtBlockID(bx & 0xF, this.py & 0xF, bz & 0xF)];
                 if (isLiquid(bb, false, null, 0, 0, 0, 0))
                   count++;
               }
@@ -479,8 +478,7 @@
           for (bz = 0; bz < this.block_side_z; bz++)
             if (this.blocks[this.py - this.yOffset][this.px][bz] != 0) {
               bb =
-                  this.ebses[this.px >> 4][bz >> 4][this.py >> 4].getBlockByExtId(this.px & 0xF,
-                      this.py & 0xF, bz & 0xF);
+                  Block.blocksList[this.ebses[this.px >> 4][bz >> 4][this.py >> 4].getExtBlockID(this.px & 0xF, this.py & 0xF, bz & 0xF)];
               meta =
                   this.ebses[this.px >> 4][bz >> 4][this.py >> 4].getExtBlockMetadata(
                       this.px & 0xF, this.py & 0xF, bz & 0xF);
@@ -502,11 +500,11 @@
         for (bx = 0; bx < this.block_side_x; bx++)
           for (bz = 0; bz < this.block_side_z; bz++)
             if ((this.blocks[this.py - this.yOffset][bx][bz] & 0x40) != 0)
-              drainBlock(bx, bz, QuarryPlusI.blockFrame);
+              drainBlock(bx, bz, QuarryPlusI.blockFrame.blockID);
       } else
         for (bz = 0; bz < this.block_side_z; bz++)
           if (this.blocks[this.py - this.yOffset][this.px][bz] != 0)
-            drainBlock(this.px, bz, Blocks.air);
+            drainBlock(this.px, bz, 0);
     S_sendNowPacket();
     return this.py < this.cy;
   }
@@ -643,10 +641,9 @@
     return false;
   }
 
-  private final void drainBlock(final int bx, final int bz, final Block tb) {
+  private final void drainBlock(final int bx, final int bz, final int tb) {
     final Block b =
-        this.ebses[bx >> 4][bz >> 4][this.py >> 4].getBlockByExtId(bx & 0xF, this.py & 0xF,
-            bz & 0xF);
+        Block.blocksList[this.ebses[bx >> 4][bz >> 4][this.py >> 4].getExtBlockID(bx & 0xF, this.py & 0xF, bz & 0xF)];
     final int meta =
         this.ebses[bx >> 4][bz >> 4][this.py >> 4].getExtBlockMetadata(bx & 0xF, this.py & 0xF,
             bz & 0xF);
--- a/com/yogpc/qp/tile/TileQuarry.java
+++ b/com/yogpc/qp/tile/TileQuarry.java
@@ -94,9 +94,9 @@
   private boolean S_checkTarget() {
     if (this.targetY > this.yMax)
       this.targetY = this.yMax;
-    final Block b =
+    final Block b =Block.blocksList[
         this.worldObj.getChunkProvider().loadChunk(this.targetX >> 4, this.targetZ >> 4)
-            .getBlock(this.targetX & 0xF, this.targetY, this.targetZ & 0xF);
+            .getBlockID(this.targetX & 0xF, this.targetY, this.targetZ & 0xF)];
     final float h =
         b == null ? -1 : b
             .getBlockHardness(this.worldObj, this.targetX, this.targetY, this.targetZ);
@@ -160,7 +160,7 @@
           return S_checkTarget();
         }
         if (b != null
-            && b.getMaterial().isSolid()
+            && this.worldObj.getBlockMaterial(this.targetX, this.targetY, this.targetZ).isSolid()
             && !(b == QuarryPlusI.blockFrame && this.worldObj.getBlockMetadata(this.targetX,
                 this.targetY, this.targetZ) == 0)) {
           this.now = NOTNEEDBREAK;
@@ -283,7 +283,7 @@
     this.digged = true;
     if (!PowerManager.useEnergyF(this, this.unbreaking))
       return false;
-    this.worldObj.setBlock(this.targetX, this.targetY, this.targetZ, QuarryPlusI.blockFrame);
+    this.worldObj.setBlock(this.targetX, this.targetY, this.targetZ, QuarryPlusI.blockFrame.blockID);
     S_setNextTarget();
     return true;
   }
@@ -358,7 +358,7 @@
 
   private boolean S_checkIAreaProvider(final int x, final int y, final int z) {
     final TileEntity te = this.worldObj.getBlockTileEntity(x, y, z);
-    if (ModAPIManager.INSTANCE.hasAPI("BuildCraftAPI|core") && te instanceof IAreaProvider) {
+    if (te instanceof IAreaProvider) {
       this.iap = (IAreaProvider) te;
       this.xMin = this.iap.xMin();
       this.xMax = this.iap.xMax();
@@ -437,7 +437,7 @@
   }
 
   private void S_setBreakableFrame(final int x, final int y, final int z) {
-    if (this.worldObj.getBlock(x, y, z) == QuarryPlusI.blockFrame)
+    if (this.worldObj.getBlockId(x, y, z) == QuarryPlusI.blockFrame.blockID)
       this.worldObj.setBlockMetadataWithNotify(x, y, z, 1, 3);
   }
 
--- a/com/yogpc/qp/tile/TileWorkbench.java
+++ b/com/yogpc/qp/tile/TileWorkbench.java
@@ -11,11 +11,9 @@
 import com.yogpc.qp.InvUtils;
 import com.yogpc.qp.WorkbenchRecipe;
 
-public class TileWorkbench extends APowerTile implements IInventory, IInventoryConnection {
+public class TileWorkbench extends APowerTile implements IInventory {
   private static final ItemStack loadItemStackFromNBT(final NBTTagCompound nbt) {
-    final Item i = Item.getItemById(nbt.getShort("id"));
-    if (i == null)
-      return null;
+    final int i = nbt.getShort("id");
     final int a = nbt.getInteger("Count");
     int d = nbt.getShort("Damage");
     if (d < 0)
@@ -27,7 +25,7 @@
   }
 
   private static final void writeItemStackToNBT(final ItemStack is, final NBTTagCompound nbt) {
-    nbt.setShort("id", (short) Item.getIdFromItem(is.getItem()));
+    nbt.setShort("id", (short) is.itemID);
     nbt.setInteger("Count", is.stackSize);
     nbt.setShort("Damage", (short) is.getItemDamage());
 
@@ -205,9 +203,4 @@
 
   @Override
   public void closeChest() {}
-
-  @Override
-    public ConnectionType canConnectInventory(final ForgeDirection arg0) {
-    return ConnectionType.FORCE;
-  }
 }
--- a/com/yogpc/qp/YogpstopPacket.java
+++ b/com/yogpc/qp/YogpstopPacket.java
@@ -90,18 +90,22 @@
     return this.data;
   }
 
-  public void readData(final ByteBuf d, final ChannelHandlerContext ctx) {
-    this.ep =
-        QuarryPlus.proxy.getPacketPlayer(ctx.channel().attr(NetworkRegistry.NET_HANDLER).get());
-    this.header = new byte[d.readInt()];
-    d.readBytes(this.header);
-    this.data = new byte[d.readableBytes()];
-    d.readBytes(this.data);
+  public void readData(final byte[] d, final Player p) {
+    this.ep = (EntityPlayer) p;
+    this.header = new byte[d[0] << 24 | (d[1] & 0xFF) << 16 | (d[2] & 0xFF) << 8 | d[3] & 0xFF];
+    this.data = new byte[d.length - this.header.length - 4];
+    System.arraycopy(d, 4, this.header, 0, this.header.length);
+    System.arraycopy(d, 4 + this.header.length, this.data, 0, this.data.length);
   }
 
-  public void writeData(final ByteBuf d) {
-    d.writeInt(this.header.length);
-    d.writeBytes(this.header);
-    d.writeBytes(this.data);
+  public byte[] writeData() {
+    final byte[] d = new byte[this.data.length + this.header.length + 4];
+    d[0] = (byte) (this.header.length >> 24);
+    d[1] = (byte) (this.header.length >> 16);
+    d[2] = (byte) (this.header.length >> 8);
+    d[3] = (byte) this.header.length;
+    System.arraycopy(this.header, 0, d, 4, this.header.length);
+    System.arraycopy(this.data, 0, d, 4 + this.header.length, this.data.length);
+    return d;
   }
 }
