--- a/com/yogpc/qp/item/ItemArmorElectric.java
+++ b/com/yogpc/qp/item/ItemArmorElectric.java
@@ -127,8 +127,8 @@
   @Override
   public void onArmorTick(final World world, final EntityPlayer player, final ItemStack is) {
     if (ElectricItem.manager != null)
-      ElectricItemManager.charge(is, ElectricItem.manager.discharge(is, Double.MAX_VALUE,
-          Integer.MAX_VALUE, true, false, false), getMaxCharge(is));
+      ElectricItemManager.charge(is, ElectricItem.manager.discharge(is, Integer.MAX_VALUE,
+          Integer.MAX_VALUE, true, false), getMaxCharge(is));
     if (player.inventory.armorInventory[2] != is)
       return;
     useJetpack(player, is, toggleHover(player, is));
@@ -214,7 +214,7 @@
 
   @Override
   public int getMaxEnergyStored(final ItemStack is) {
-    return (int) (getMaxCharge(is) * 4);
+    return getMaxCharge(is) * 4;
   }
 
   @Override
--- a/com/yogpc/qp/tile/TilePump.java
+++ b/com/yogpc/qp/tile/TilePump.java
@@ -531,7 +531,7 @@
         final FluidStack s = this.liquids.get(i);
         ret[i + 1] =
             new ChatComponentTranslation("yog.pump.liquid", new ChatComponentTranslation(s
-                .getFluid().getUnlocalizedName(s)), Integer.toString(s.amount));
+                .getFluid().getUnlocalizedName()), Integer.toString(s.amount));
       }
     } else
       ret[0] = new ChatComponentTranslation("chat.pumpcontainno");
--- a/com/yogpc/qp/gui/GuiP_List.java
+++ b/com/yogpc/qp/gui/GuiP_List.java
@@ -94,8 +94,7 @@
       case PacketHandler.CtS_REMOVE_MAPPING:
         String name = this.tile.mapping[this.dir].get(this.oreslot.currentore);
         if (FluidRegistry.isFluidRegistered(name))
-          name =
-              FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+          name = FluidRegistry.getFluid(name).getLocalizedName();
         this.mc.displayGuiScreen(new GuiYesNo(this, StatCollector
             .translateToLocal("tof.deletefluidsure"), name, par1.id));
         break;
--- a/com/yogpc/qp/gui/GuiP_Manual.java
+++ b/com/yogpc/qp/gui/GuiP_Manual.java
@@ -57,15 +57,13 @@
           return;
         if (this.tile.mapping[this.targetid].contains(name)) {
           if (FluidRegistry.isFluidRegistered(name))
-            name =
-                FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+            name = FluidRegistry.getFluid(name).getLocalizedName();
           this.mc.displayGuiScreen(new GuiError(this, StatCollector
               .translateToLocal("tof.alreadyerror"), name));
           return;
         }
         if (FluidRegistry.isFluidRegistered(name))
-          name =
-              FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+          name = FluidRegistry.getFluid(name).getLocalizedName();
         this.mc.displayGuiScreen(new GuiYesNo(this, StatCollector
             .translateToLocal("tof.addfluidsure"), name, -1));
         break;
--- a/com/yogpc/qp/gui/GuiP_SlotBlockList.java
+++ b/com/yogpc/qp/gui/GuiP_SlotBlockList.java
@@ -74,9 +74,7 @@
   @Override
   protected void drawSlot(final int i, final int v2, final int v3, final int v4,
       final Tessellator t, final int v6, final int v7) {
-    final String name =
-        FluidRegistry.getFluid(this.blocklist.get(i)).getLocalizedName(
-            FluidRegistry.getFluidStack(this.blocklist.get(i), 0));
+    final String name = FluidRegistry.getFluid(this.blocklist.get(i)).getLocalizedName();
     Minecraft.getMinecraft().fontRenderer.drawStringWithShadow(name,
         (this.parent.width - Minecraft.getMinecraft().fontRenderer.getStringWidth(name)) / 2,
         v3 + 2, 0xFFFFFF);
--- a/com/yogpc/qp/gui/GuiP_SlotList.java
+++ b/com/yogpc/qp/gui/GuiP_SlotList.java
@@ -61,7 +61,7 @@
       final Tessellator t, final int v6, final int v7) {
     String name = this.target.get(i);
     if (FluidRegistry.isFluidRegistered(name))
-      name = FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+      name = FluidRegistry.getFluid(name).getLocalizedName();
     Minecraft.getMinecraft().fontRenderer
         .drawStringWithShadow(
             name,
