--- 1.7.2a/java/com/yogpc/qp/PacketHandler.java
+++ 1.7.2/java/com/yogpc/qp/PacketHandler.java
@@ -29,7 +29,6 @@
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.inventory.Container;
 import net.minecraft.nbt.CompressedStreamTools;
-import net.minecraft.nbt.NBTSizeTracker;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.tileentity.TileEntity;
 
@@ -139,7 +138,7 @@
 	private static void setNBTFromPacket(QuarryPlusPacket p) {
 		try {
 			NBTTagCompound cache;
-			cache = CompressedStreamTools.func_152457_a(p.getData(), NBTSizeTracker.field_152451_a);
+			cache = CompressedStreamTools.decompress(p.getData());
 			TileEntity te = p.getPlayer().worldObj.getTileEntity(cache.getInteger("x"), cache.getInteger("y"), cache.getInteger("z"));
 			if (te != null) te.readFromNBT(cache);
 		} catch (IOException e) {
--- 1.7.2a/java/com/yogpc/qp/TilePump.java
+++ 1.7.2/java/com/yogpc/qp/TilePump.java
@@ -488,7 +488,7 @@
 		if (this.liquids.size() > 0) {
 			ret[0] = StatCollector.translateToLocal("chat.pumpcontain");
 			for (int i = 0; i < this.liquids.size(); i++) {
-				ret[i + 1] = new StringBuilder().append("    ").append(this.liquids.get(i).getFluid().getLocalizedName(this.liquids.get(i))).append(": ")
+				ret[i + 1] = new StringBuilder().append("    ").append(this.liquids.get(i).getFluid().getLocalizedName()).append(": ")
 						.append(this.liquids.get(i).amount).append("mB").toString();
 			}
 		} else {
--- 1.7.2a/java/com/yogpc/qp/client/GuiP_List.java
+++ 1.7.2/java/com/yogpc/qp/client/GuiP_List.java
@@ -23,13 +23,12 @@
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import net.minecraft.client.gui.GuiButton;
-import net.minecraft.client.gui.GuiYesNoCallback;
 import net.minecraft.util.StatCollector;
 import net.minecraftforge.common.util.ForgeDirection;
 import net.minecraftforge.fluids.FluidRegistry;
 
 @SideOnly(Side.CLIENT)
-public class GuiP_List extends GuiScreenA implements GuiYesNoCallback {
+public class GuiP_List extends GuiScreenA {
 	private GuiP_SlotList oreslot;
 	private GuiButton delete, top, up, down, bottom;
 	private TilePump tile;
@@ -81,7 +80,7 @@
 			break;
 		case PacketHandler.CtS_REMOVE_MAPPING:
 			String name = this.tile.mapping[this.dir].get(this.oreslot.currentore);
-			if (FluidRegistry.isFluidRegistered(name)) name = FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+			if (FluidRegistry.isFluidRegistered(name)) name = FluidRegistry.getFluid(name).getLocalizedName();
 			this.mc.displayGuiScreen(new GuiYesNo(this, StatCollector.translateToLocal("tof.deletefluidsure"), name, par1.id));
 			break;
 		default:
--- 1.7.2a/java/com/yogpc/qp/client/GuiP_Manual.java
+++ 1.7.2/java/com/yogpc/qp/client/GuiP_Manual.java
@@ -25,12 +25,11 @@
 import net.minecraft.client.gui.GuiButton;
 import net.minecraft.client.gui.GuiScreen;
 import net.minecraft.client.gui.GuiTextField;
-import net.minecraft.client.gui.GuiYesNoCallback;
 import net.minecraft.util.StatCollector;
 import net.minecraftforge.fluids.FluidRegistry;
 
 @SideOnly(Side.CLIENT)
-public class GuiP_Manual extends GuiScreenA implements GuiYesNoCallback {
+public class GuiP_Manual extends GuiScreenA {
 	private GuiTextField blockid;
 	private byte targetid;
 	private TilePump tile;
@@ -56,11 +55,11 @@
 			String name = this.blockid.getText();
 			if (name.length() == 0) return;
 			if (this.tile.mapping[this.targetid].contains(name)) {
-				if (FluidRegistry.isFluidRegistered(name)) name = FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+				if (FluidRegistry.isFluidRegistered(name)) name = FluidRegistry.getFluid(name).getLocalizedName();
 				this.mc.displayGuiScreen(new GuiError(this, StatCollector.translateToLocal("tof.alreadyerror"), name));
 				return;
 			}
-			if (FluidRegistry.isFluidRegistered(name)) name = FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+			if (FluidRegistry.isFluidRegistered(name)) name = FluidRegistry.getFluid(name).getLocalizedName();
 			this.mc.displayGuiScreen(new GuiYesNo(this, StatCollector.translateToLocal("tof.addfluidsure"), name, -1));
 			break;
 		case -2:
--- 1.7.2a/java/com/yogpc/qp/client/GuiP_SlotBlockList.java
+++ 1.7.2/java/com/yogpc/qp/client/GuiP_SlotBlockList.java
@@ -77,7 +77,7 @@
 
 	@Override
 	protected void drawSlot(int var1, int var2, int var3, int var4, Tessellator var5, int var6, int var7) {
-		String name = FluidRegistry.getFluid(this.blocklist.get(var1)).getLocalizedName(FluidRegistry.getFluidStack(this.blocklist.get(var1), 0));
+		String name = FluidRegistry.getFluid(this.blocklist.get(var1)).getLocalizedName();
 		Minecraft.getMinecraft().fontRenderer.drawStringWithShadow(name, (this.parent.width - Minecraft.getMinecraft().fontRenderer.getStringWidth(name)) / 2,
 				var3 + 1, 0xFFFFFF);
 	}
--- 1.7.2a/java/com/yogpc/qp/client/GuiP_SlotList.java
+++ 1.7.2/java/com/yogpc/qp/client/GuiP_SlotList.java
@@ -62,7 +62,7 @@
 	@Override
 	protected void drawSlot(int var1, int var2, int var3, int var4, Tessellator var5, int var6, int var7) {
 		String name = this.target.get(var1);
-		if (FluidRegistry.isFluidRegistered(name)) name = FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+		if (FluidRegistry.isFluidRegistered(name)) name = FluidRegistry.getFluid(name).getLocalizedName();
 		Minecraft.getMinecraft().fontRenderer.drawStringWithShadow(name,
 				(this.parent.width * 3 / 5 - Minecraft.getMinecraft().fontRenderer.getStringWidth(name)) / 2, var3 + 1, 0xFFFFFF);
 	}
--- 1.7.2a/java/com/yogpc/qp/client/GuiQ_List.java
+++ 1.7.2/java/com/yogpc/qp/client/GuiQ_List.java
@@ -24,11 +24,10 @@
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import net.minecraft.client.gui.GuiButton;
-import net.minecraft.client.gui.GuiYesNoCallback;
 import net.minecraft.util.StatCollector;
 
 @SideOnly(Side.CLIENT)
-public class GuiQ_List extends GuiScreenA implements GuiYesNoCallback {
+public class GuiQ_List extends GuiScreenA {
 	private GuiQ_SlotList oreslot;
 	private GuiButton delete;
 	private TileBasic tile;
--- 1.7.2a/java/com/yogpc/qp/client/GuiYesNo.java
+++ 1.7.2/java/com/yogpc/qp/client/GuiYesNo.java
@@ -19,16 +19,16 @@
 
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
-import net.minecraft.client.gui.GuiYesNoCallback;
+import net.minecraft.client.gui.GuiScreen;
 
 @SideOnly(Side.CLIENT)
 public class GuiYesNo extends net.minecraft.client.gui.GuiYesNo {
 
-	public GuiYesNo(GuiYesNoCallback par1GuiScreen, String par2Str, String par3Str, int par4) {
+	public GuiYesNo(GuiScreen par1GuiScreen, String par2Str, String par3Str, int par4) {
 		super(par1GuiScreen, par2Str, par3Str, par4);
 	}
 
-	public GuiYesNo(GuiYesNoCallback par1GuiScreen, String par2Str, String par3Str, String par4Str, String par5Str, int par6) {
+	public GuiYesNo(GuiScreen par1GuiScreen, String par2Str, String par3Str, String par4Str, String par5Str, int par6) {
 		super(par1GuiScreen, par2Str, par3Str, par4Str, par5Str, par6);
 	}
 
