--- a/com/yogpc/ip/ItemArmorElectric.java
+++ b/com/yogpc/ip/ItemArmorElectric.java
@@ -127,8 +127,8 @@
   @Override
   public void onArmorTick(final World world, final EntityPlayer player, final ItemStack is) {
     if (ElectricItem.manager != null)
-      ElectricItemManager.charge(is, ElectricItem.manager.discharge(is, Double.MAX_VALUE,
-          Integer.MAX_VALUE, true, false, false), getMaxCharge(is));
+      ElectricItemManager.charge(is, ElectricItem.manager.discharge(is, Integer.MAX_VALUE,
+          Integer.MAX_VALUE, true, false), getMaxCharge(is));
     if (player.inventory.armorInventory[2] != is)
       return;
     useJetpack(player, is, toggleHover(player, is));
@@ -214,7 +214,7 @@
 
   @Override
   public int getMaxEnergyStored(final ItemStack is) {
-    return (int) (getMaxCharge(is) * 4);
+    return getMaxCharge(is) * 4;
   }
 
   @Override
--- a/com/yogpc/mc_lib/ContainerWorkbench.java
+++ b/com/yogpc/mc_lib/ContainerWorkbench.java
@@ -31,7 +31,7 @@
     if (nbtlen >= 0) {
       final byte[] ba = new byte[nbtlen];
       bb.readFully(ba);
-      is.setTagCompound(CompressedStreamTools.func_152457_a(ba, NBTSizeTracker.field_152451_a));
+      is.setTagCompound(CompressedStreamTools.decompress(ba));
     }
     return is;
   }
--- a/com/yogpc/mc_lib/PacketHandler.java
+++ b/com/yogpc/mc_lib/PacketHandler.java
@@ -150,8 +150,7 @@
 
   private static void setNBTFromPacket(final YogpstopPacket p) {
     try {
-      final NBTTagCompound cache =
-          CompressedStreamTools.func_152457_a(p.getData(), NBTSizeTracker.field_152451_a);
+      final NBTTagCompound cache = CompressedStreamTools.decompress(p.getData());
       final TileEntity te =
           p.getPlayer().worldObj.getTileEntity(cache.getInteger("x"), cache.getInteger("y"),
               cache.getInteger("z"));
--- a/com/yogpc/qp/TilePump.java
+++ b/com/yogpc/qp/TilePump.java
@@ -531,7 +531,7 @@
         final FluidStack s = this.liquids.get(i);
         ret[i + 1] =
             new ChatComponentTranslation("yog.pump.liquid", new ChatComponentTranslation(s
-                .getFluid().getUnlocalizedName(s)), Integer.toString(s.amount));
+                .getFluid().getUnlocalizedName()), Integer.toString(s.amount));
       }
     } else
       ret[0] = new ChatComponentTranslation("chat.pumpcontainno");
--- a/com/yogpc/qp/client/GuiEnchList.java
+++ b/com/yogpc/qp/client/GuiEnchList.java
@@ -32,7 +32,7 @@
 import cpw.mods.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class GuiEnchList extends GuiContainer implements GuiYesNoCallback {
+public class GuiEnchList extends GuiContainer {
   private GuiSlotEnchList slot;
   private final TileBasic tile;
   private final byte targetid;
--- a/com/yogpc/qp/client/GuiP_List.java
+++ b/com/yogpc/qp/client/GuiP_List.java
@@ -25,7 +25,7 @@
 import cpw.mods.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class GuiP_List extends GuiScreenA implements GuiYesNoCallback {
+public class GuiP_List extends GuiScreenA {
   private GuiP_SlotList oreslot;
   private GuiButton delete, top, up, down, bottom;
   private final TilePump tile;
@@ -94,8 +94,7 @@
       case PacketHandler.CtS_REMOVE_MAPPING:
         String name = this.tile.mapping[this.dir].get(this.oreslot.currentore);
         if (FluidRegistry.isFluidRegistered(name))
-          name =
-              FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+          name = FluidRegistry.getFluid(name).getLocalizedName();
         this.mc.displayGuiScreen(new GuiYesNo(this, StatCollector
             .translateToLocal("tof.deletefluidsure"), name, par1.id));
         break;
--- a/com/yogpc/qp/client/GuiP_Manual.java
+++ b/com/yogpc/qp/client/GuiP_Manual.java
@@ -26,7 +26,7 @@
 import cpw.mods.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class GuiP_Manual extends GuiScreenA implements GuiYesNoCallback {
+public class GuiP_Manual extends GuiScreenA {
   private GuiTextField blockid;
   private final byte targetid;
   private final TilePump tile;
@@ -57,15 +57,13 @@
           return;
         if (this.tile.mapping[this.targetid].contains(name)) {
           if (FluidRegistry.isFluidRegistered(name))
-            name =
-                FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+            name = FluidRegistry.getFluid(name).getLocalizedName();
           this.mc.displayGuiScreen(new GuiError(this, StatCollector
               .translateToLocal("tof.alreadyerror"), name));
           return;
         }
         if (FluidRegistry.isFluidRegistered(name))
-          name =
-              FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+          name = FluidRegistry.getFluid(name).getLocalizedName();
         this.mc.displayGuiScreen(new GuiYesNo(this, StatCollector
             .translateToLocal("tof.addfluidsure"), name, -1));
         break;
--- a/com/yogpc/qp/client/GuiP_SlotBlockList.java
+++ b/com/yogpc/qp/client/GuiP_SlotBlockList.java
@@ -74,9 +74,7 @@
   @Override
   protected void drawSlot(final int i, final int v2, final int v3, final int v4,
       final Tessellator t, final int v6, final int v7) {
-    final String name =
-        FluidRegistry.getFluid(this.blocklist.get(i)).getLocalizedName(
-            FluidRegistry.getFluidStack(this.blocklist.get(i), 0));
+    final String name = FluidRegistry.getFluid(this.blocklist.get(i)).getLocalizedName();
     Minecraft.getMinecraft().fontRenderer.drawStringWithShadow(name,
         (this.parent.width - Minecraft.getMinecraft().fontRenderer.getStringWidth(name)) / 2,
         v3 + 2, 0xFFFFFF);
--- a/com/yogpc/qp/client/GuiP_SlotList.java
+++ b/com/yogpc/qp/client/GuiP_SlotList.java
@@ -61,7 +61,7 @@
       final Tessellator t, final int v6, final int v7) {
     String name = this.target.get(i);
     if (FluidRegistry.isFluidRegistered(name))
-      name = FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+      name = FluidRegistry.getFluid(name).getLocalizedName();
     Minecraft.getMinecraft().fontRenderer
         .drawStringWithShadow(
             name,
--- a/com/yogpc/qp/client/GuiYesNo.java
+++ b/com/yogpc/qp/client/GuiYesNo.java
@@ -19,12 +19,12 @@
 @SideOnly(Side.CLIENT)
 public class GuiYesNo extends net.minecraft.client.gui.GuiYesNo {
 
-  public GuiYesNo(final GuiYesNoCallback par1GuiScreen, final String par2Str, final String par3Str,
+  public GuiYesNo(final GuiScreen par1GuiScreen, final String par2Str, final String par3Str,
       final int par4) {
     super(par1GuiScreen, par2Str, par3Str, par4);
   }
 
-  public GuiYesNo(final GuiYesNoCallback par1GuiScreen, final String par2Str, final String par3Str,
+  public GuiYesNo(final GuiScreen par1GuiScreen, final String par2Str, final String par3Str,
       final String par4Str, final String par5Str, final int par6) {
     super(par1GuiScreen, par2Str, par3Str, par4Str, par5Str, par6);
   }
