--- a/com/yogpc/ip/ItemArmorElectric.java
+++ b/com/yogpc/ip/ItemArmorElectric.java
@@ -80,7 +80,7 @@
           }
         }
       }
-      ElectricItem.manager.discharge(jetpack, hover ? 7 : 8, Integer.MAX_VALUE, true, false, false);
+      ElectricItem.manager.discharge(jetpack, hover ? 7 : 8, Integer.MAX_VALUE, true, false);
       {
         p.fallDistance = 0.0f;
         p.distanceWalkedModified = 0.0f;
@@ -139,7 +139,7 @@
   @SideOnly(Side.CLIENT)
   public void getSubItems(final Item item, final CreativeTabs par2CreativeTabs, final List itemList) {
     final ItemStack charged = new ItemStack(this, 1);
-    ElectricItem.manager.charge(charged, Double.POSITIVE_INFINITY, Integer.MAX_VALUE, true, false);
+    ElectricItem.manager.charge(charged, Integer.MAX_VALUE, Integer.MAX_VALUE, true, false);
     itemList.add(charged);
     itemList.add(new ItemStack(this, 1, getMaxDamage()));
   }
@@ -164,7 +164,7 @@
   @Override
   public void damageArmor(final EntityLivingBase entity, final ItemStack stack,
       final DamageSource source, final int damage, final int slot) {
-    ElectricItem.manager.discharge(stack, damage * 5000, Integer.MAX_VALUE, true, false, false);
+    ElectricItem.manager.discharge(stack, damage * 5000, Integer.MAX_VALUE, true, false);
   }
 
   @Override
@@ -214,7 +214,7 @@
 
   @Override
   public int receiveEnergy(final ItemStack is, final int am, final boolean sim) {
-    return (int) (ElectricItem.manager.charge(is, (double) am / 4, Integer.MAX_VALUE, false, false) * 4);
+    return (int) (ElectricItem.manager.charge(is, (int) ((double) am / 4), Integer.MAX_VALUE, false, false) * 4);
   }
 
 }
--- a/com/yogpc/mc_lib/PacketHandler.java
+++ b/com/yogpc/mc_lib/PacketHandler.java
@@ -33,7 +33,6 @@
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.inventory.Container;
 import net.minecraft.nbt.CompressedStreamTools;
-import net.minecraft.nbt.NBTSizeTracker;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.tileentity.TileEntity;
 
@@ -160,8 +159,7 @@
 
   private static void setNBTFromPacket(final YogpstopPacket p) {
     try {
-      final NBTTagCompound cache =
-          CompressedStreamTools.func_152457_a(p.getData(), NBTSizeTracker.field_152451_a);
+      final NBTTagCompound cache = CompressedStreamTools.decompress(p.getData());
       final TileEntity te =
           p.getPlayer().worldObj.getTileEntity(cache.getInteger("x"), cache.getInteger("y"),
               cache.getInteger("z"));
--- a/com/yogpc/qp/TilePump.java
+++ b/com/yogpc/qp/TilePump.java
@@ -526,8 +526,8 @@
       for (int i = 0; i < this.liquids.size(); i++)
         ret[i + 1] =
             new StringBuilder().append("    ")
-                .append(this.liquids.get(i).getFluid().getLocalizedName(this.liquids.get(i)))
-                .append(": ").append(this.liquids.get(i).amount).append("mB").toString();
+                .append(this.liquids.get(i).getFluid().getLocalizedName()).append(": ")
+                .append(this.liquids.get(i).amount).append("mB").toString();
     } else
       ret[0] = StatCollector.translateToLocal("chat.pumpcontainno");
     return ret;
--- a/com/yogpc/qp/client/GuiP_List.java
+++ b/com/yogpc/qp/client/GuiP_List.java
@@ -14,7 +14,6 @@
 package com.yogpc.qp.client;
 
 import net.minecraft.client.gui.GuiButton;
-import net.minecraft.client.gui.GuiYesNoCallback;
 import net.minecraft.util.StatCollector;
 import net.minecraftforge.common.util.ForgeDirection;
 import net.minecraftforge.fluids.FluidRegistry;
@@ -26,7 +25,7 @@
 import cpw.mods.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class GuiP_List extends GuiScreenA implements GuiYesNoCallback {
+public class GuiP_List extends GuiScreenA {
   private GuiP_SlotList oreslot;
   private GuiButton delete, top, up, down, bottom;
   private final TilePump tile;
@@ -94,8 +93,7 @@
       case PacketHandler.CtS_REMOVE_MAPPING:
         String name = this.tile.mapping[this.dir].get(this.oreslot.currentore);
         if (FluidRegistry.isFluidRegistered(name))
-          name =
-              FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+          name = FluidRegistry.getFluid(name).getLocalizedName();
         this.mc.displayGuiScreen(new GuiYesNo(this, StatCollector
             .translateToLocal("tof.deletefluidsure"), name, par1.id));
         break;
--- a/com/yogpc/qp/client/GuiP_Manual.java
+++ b/com/yogpc/qp/client/GuiP_Manual.java
@@ -16,7 +16,6 @@
 import net.minecraft.client.gui.GuiButton;
 import net.minecraft.client.gui.GuiScreen;
 import net.minecraft.client.gui.GuiTextField;
-import net.minecraft.client.gui.GuiYesNoCallback;
 import net.minecraft.util.StatCollector;
 import net.minecraftforge.fluids.FluidRegistry;
 
@@ -27,7 +26,7 @@
 import cpw.mods.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class GuiP_Manual extends GuiScreenA implements GuiYesNoCallback {
+public class GuiP_Manual extends GuiScreenA {
   private GuiTextField blockid;
   private final byte targetid;
   private final TilePump tile;
@@ -57,15 +56,13 @@
           return;
         if (this.tile.mapping[this.targetid].contains(name)) {
           if (FluidRegistry.isFluidRegistered(name))
-            name =
-                FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+            name = FluidRegistry.getFluid(name).getLocalizedName();
           this.mc.displayGuiScreen(new GuiError(this, StatCollector
               .translateToLocal("tof.alreadyerror"), name));
           return;
         }
         if (FluidRegistry.isFluidRegistered(name))
-          name =
-              FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+          name = FluidRegistry.getFluid(name).getLocalizedName();
         this.mc.displayGuiScreen(new GuiYesNo(this, StatCollector
             .translateToLocal("tof.addfluidsure"), name, -1));
         break;
--- a/com/yogpc/qp/client/GuiP_SlotBlockList.java
+++ b/com/yogpc/qp/client/GuiP_SlotBlockList.java
@@ -75,9 +75,7 @@
   @Override
   protected void drawSlot(final int var1, final int var2, final int var3, final int var4,
       final Tessellator var5, final int var6, final int var7) {
-    final String name =
-        FluidRegistry.getFluid(this.blocklist.get(var1)).getLocalizedName(
-            FluidRegistry.getFluidStack(this.blocklist.get(var1), 0));
+    final String name = FluidRegistry.getFluid(this.blocklist.get(var1)).getLocalizedName();
     Minecraft.getMinecraft().fontRenderer.drawStringWithShadow(name,
         (this.parent.width - Minecraft.getMinecraft().fontRenderer.getStringWidth(name)) / 2,
         var3 + 1, 0xFFFFFF);
--- a/com/yogpc/qp/client/GuiP_SlotList.java
+++ b/com/yogpc/qp/client/GuiP_SlotList.java
@@ -62,7 +62,7 @@
       final Tessellator var5, final int var6, final int var7) {
     String name = this.target.get(var1);
     if (FluidRegistry.isFluidRegistered(name))
-      name = FluidRegistry.getFluid(name).getLocalizedName(FluidRegistry.getFluidStack(name, 0));
+      name = FluidRegistry.getFluid(name).getLocalizedName();
     Minecraft.getMinecraft().fontRenderer
         .drawStringWithShadow(
             name,
--- a/com/yogpc/qp/client/GuiQ_List.java
+++ b/com/yogpc/qp/client/GuiQ_List.java
@@ -17,7 +17,6 @@
 import java.io.DataOutputStream;
 
 import net.minecraft.client.gui.GuiButton;
-import net.minecraft.client.gui.GuiYesNoCallback;
 import net.minecraft.util.StatCollector;
 
 import com.yogpc.mc_lib.PacketHandler;
@@ -30,7 +29,7 @@
 import cpw.mods.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class GuiQ_List extends GuiScreenA implements GuiYesNoCallback {
+public class GuiQ_List extends GuiScreenA {
   private GuiQ_SlotList oreslot;
   private GuiButton delete;
   private final TileBasic tile;
--- a/com/yogpc/qp/client/GuiYesNo.java
+++ b/com/yogpc/qp/client/GuiYesNo.java
@@ -13,19 +13,19 @@
 
 package com.yogpc.qp.client;
 
-import net.minecraft.client.gui.GuiYesNoCallback;
+import net.minecraft.client.gui.GuiScreen;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
 public class GuiYesNo extends net.minecraft.client.gui.GuiYesNo {
 
-  public GuiYesNo(final GuiYesNoCallback par1GuiScreen, final String par2Str, final String par3Str,
+  public GuiYesNo(final GuiScreen par1GuiScreen, final String par2Str, final String par3Str,
       final int par4) {
     super(par1GuiScreen, par2Str, par3Str, par4);
   }
 
-  public GuiYesNo(final GuiYesNoCallback par1GuiScreen, final String par2Str, final String par3Str,
+  public GuiYesNo(final GuiScreen par1GuiScreen, final String par2Str, final String par3Str,
       final String par4Str, final String par5Str, final int par6) {
     super(par1GuiScreen, par2Str, par3Str, par4Str, par5Str, par6);
   }
