--- a/com/yogpc/ip/ItemArmorElectric.java
+++ b/com/yogpc/ip/ItemArmorElectric.java
@@ -31,7 +31,7 @@
 import cpw.mods.fml.common.asm.SideOnly;
 
 public class ItemArmorElectric extends ItemArmor implements ISpecialArmor, IElectricItem,
-    IArmorTextureProvider {
+    IArmorTextureProvider, ITickHandler {
   public ItemArmorElectric(final int id) {
     super(id, EnumHelper.addArmorMaterial("ELECTRIC", 33, new int[] {3, 8, 6, 3}, 0), YogpstopLib.proxy
         .addNewArmourRendererPrefix("electric"), 1);
@@ -40,6 +40,8 @@
     setMaxDamage(100);
     setMaxStackSize(1);
     this.iconIndex = 83;
+    TickRegistry.registerTickHandler(this, Side.CLIENT);
+    TickRegistry.registerTickHandler(this, Side.SERVER);
   }
 
   @Override
@@ -122,16 +124,31 @@
   }
 
   @Override
-  public void onArmorTickUpdate(final World world, final EntityPlayer player, final ItemStack is) {
+  public void tickStart(final EnumSet<TickType> type, final Object... tickData) {
+    final EntityPlayer player = (EntityPlayer) tickData[0];
+    final ItemStack is = player.inventory.armorInventory[2];
     if (Loader.isModLoaded("IC2"))
       ElectricItemManager.charge(is, ElectricItem.discharge(is, Integer.MAX_VALUE,
           Integer.MAX_VALUE, true, false), getMaxCharge());
-    if (player.inventory.armorInventory[2] != is)
+    if (this != is.getItem())
       return;
     useJetpack(player, is, toggleHover(player, is));
   }
 
   @Override
+  public void tickEnd(final EnumSet<TickType> type, final Object... tickData) {}
+
+  @Override
+  public EnumSet<TickType> ticks() {
+    return EnumSet.of(TickType.PLAYER);
+  }
+
+  @Override
+  public String getLabel() {
+    return "JetPack";
+  }
+
+  @Override
   @SideOnly(Side.CLIENT)
   public EnumRarity getRarity(final ItemStack stack) {
     return EnumRarity.uncommon;
--- a/com/yogpc/mc_lib/APowerTile.java
+++ b/com/yogpc/mc_lib/APowerTile.java
@@ -99,7 +99,7 @@
 
   @Override
   public boolean demandsEnergy() {
-    return (int) (Math.min(this.maxGot - this.got, this.max - this.all - this.got) * 2.5);
+    return Math.min(this.maxGot - this.got, this.max - this.all - this.got) * 2.5 > 0;
   }
 
   @Override
@@ -108,11 +108,6 @@
   }
 
   @Override
-  public int getMaxSafeInput() {
-    return 3;
-  }
-
-  @Override
   public boolean isAddedToEnergyNet() {
     return true;
   }
--- a/com/yogpc/mc_lib/ContainerWorkbench.java
+++ b/com/yogpc/mc_lib/ContainerWorkbench.java
@@ -39,7 +39,7 @@
       bb.writeShort((short) -1);
       return;
     }
-    bb.writeShort((short) is.shiftedIndex);
+    bb.writeShort((short) is.itemID);
     bb.writeInt(is.stackSize);
     bb.writeShort((short) is.getItemDamage());
     NBTTagCompound nbttagcompound = null;
--- a/com/yogpc/mc_lib/TileWorkbench.java
+++ b/com/yogpc/mc_lib/TileWorkbench.java
@@ -21,7 +21,7 @@
   }
 
   private static final void writeItemStackToNBT(final ItemStack is, final NBTTagCompound nbt) {
-    nbt.setShort("id", (short) is.shiftedIndex);
+    nbt.setShort("id", (short) is.itemID);
     nbt.setInteger("Count", is.stackSize);
     nbt.setShort("Damage", (short) is.getItemDamage());
 
--- a/com/yogpc/qp/client/RenderRefinery.java
+++ b/com/yogpc/qp/client/RenderRefinery.java
@@ -177,7 +177,7 @@
           GL11.glPushMatrix();
           GL11.glTranslatef(tx, ty, tz);
         }
-        bindTextureByName(Block.blocksList[liquid.shiftedIndex].getTextureFile());
+        bindTextureByName(Block.blocksList[liquid.itemID].getTextureFile());
         final ItemStack stack = liquid.asItemStack();
         setColor(stack.getItem().getColorFromItemStack(stack, 0));
         GL11.glCallList(list[(int) (liquid.amount / buf * 99)]);
@@ -196,8 +196,8 @@
       return diplayLists;
     diplayLists = new int[100];
     Block baseBlock;
-    if (fluidStack.shiftedIndex > -1) {
-      baseBlock = Block.blocksList[fluidStack.shiftedIndex];
+    if (fluidStack.itemID > -1) {
+      baseBlock = Block.blocksList[fluidStack.itemID];
     } else {
       baseBlock = Block.waterStill;
     }
