--- a/com/yogpc/ip/BlockController.java
+++ b/com/yogpc/ip/BlockController.java
@@ -37,12 +37,13 @@
     setBlockName("spawnercontroller");
     setHardness(1.0f);
     setCreativeTab(CreativeTabs.tabRedstone);
+    this.blockIndexInTexture = 85;
   }
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister ir) {
-    this.blockIndexInTexture = ir.registerIcon("yogpstop_qp:spawnctl");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   private static final TileEntityMobSpawner getSpawner(final World w, final int x, final int y,
@@ -51,10 +52,7 @@
       final TileEntity t = w.getBlockTileEntity(x + d.offsetX, y + d.offsetY, z + d.offsetZ);
       if (!(t instanceof TileEntityMobSpawner))
         continue;
-      final TileEntityMobSpawner l = ((TileEntityMobSpawner) t).func_98049_a();
-      if (l == null)
-        continue;
-      return l;
+      return (TileEntityMobSpawner) t;
     }
     return null;
   }
@@ -124,13 +122,11 @@
     if (r && m == 0) {
       final TileEntityMobSpawner l = getSpawner(w, x, y, z);
       if (l != null) {
-        l.spawnDelay = 0;
-        final EntityPlayer p = CommonProxy.getPlayer(w);
-        p.setWorld(l.worldObj);
-        p.setPosition(l.xCoord, l.yCoord, l.zCoord);
-        l.worldObj.playerEntities.add(p);
-        l.updateSpawner();
-        l.worldObj.playerEntities.remove(p);
+        l.delay = 0;
+        final EntityPlayer p = CommonProxy.getPlayer(w,x,y,z);
+        l.worldObj.playerEntities.add(p);
+        l.updateEntity();
+        l.worldObj.playerEntities.remove(p);
       }
       w.setBlockMetadataWithNotify(x, y, z, 1);
     } else if (!r && m == 1)
--- a/com/yogpc/ip/ItemArmorElectric.java
+++ b/com/yogpc/ip/ItemArmorElectric.java
@@ -26,25 +26,27 @@
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 
-public class ItemArmorElectric extends ItemArmor implements ISpecialArmor, IElectricItem {
+public class ItemArmorElectric extends ItemArmor implements ISpecialArmor, IElectricItem,
+    IArmorTextureProvider {
   public ItemArmorElectric(final int id) {
     super(id, EnumHelper.addArmorMaterial("ELECTRIC", 33, new int[] {3, 8, 6, 3}, 0), YogpstopLib.proxy
         .addNewArmourRendererPrefix("electric"), 1);
-    setBlockName("electric_armor");
+    setItemName("electric_armor");
     setCreativeTab(CreativeTabs.tabCombat);
     setMaxDamage(100);
     setMaxStackSize(1);
+    this.iconIndex = 83;
   }
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister ir) {
-    this.iconIndex = ir.registerIcon("yogpstop_qp:elecArmor");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
-  public String getArmorTexture(ItemStack i, Entity e, int s, int l) {
-    return "yogpstop_qp:textures/models/armor/elecArmor_layer_1.png";
+  public String getArmorTextureFile(ItemStack i) {
+    return "/mods/yogpstop_qp/textures/models/armor/elecArmor_layer_1.png";
   }
 
   private static final double maxFallSpeedOnHover = 0.35;
@@ -117,7 +119,6 @@
 
   @Override
   public void onArmorTickUpdate(final World world, final EntityPlayer player, final ItemStack is) {
-    if (ElectricItem != null)
       ElectricItemManager.charge(is, ElectricItem.discharge(is, Integer.MAX_VALUE,
           Integer.MAX_VALUE, true, false), getMaxCharge());
     if (player.inventory.armorInventory[2] != is)
--- a/com/yogpc/ip/ItemMirror.java
+++ b/com/yogpc/ip/ItemMirror.java
@@ -16,14 +16,15 @@
 public class ItemMirror extends Item {
   public ItemMirror(final int id) {
     super(id);
-    setBlockName("magicmirror");
+    setItemName("magicmirror");
     setCreativeTab(CreativeTabs.tabTools);
+    this.iconIndex = 84;
   }
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister ir) {
-    this.iconIndex = ir.registerIcon("yogpstop_qp:mm");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
--- a/com/yogpc/mc_lib/BlockWorkbench.java
+++ b/com/yogpc/mc_lib/BlockWorkbench.java
@@ -21,12 +21,13 @@
     setHardness(3F);
     setCreativeTab(CreativeTabs.tabRedstone);
     setBlockName("WorkbenchPlus");
+    this.blockIndexInTexture = 37;
   }
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister ir) {
-    this.blockIndexInTexture = ir.registerIcon("yogpstop_qp:workbench");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
@@ -57,7 +58,6 @@
           }
         }
       }
-      w.func_96440_m(x, y, z, b);
     }
     super.breakBlock(w, x, y, z, b, m);
   }
--- a/com/yogpc/mc_lib/ContainerWorkbench.java
+++ b/com/yogpc/mc_lib/ContainerWorkbench.java
@@ -184,19 +184,6 @@
     return;
   }
 
-  private int field_94535_f = -1;
-  /*
-   * drag_state 0 start 1 content 2 end
-   */
-  private int drag_state;
-  private final Set<Slot> dragged = new HashSet<Slot>();
-
-  @Override
-  protected void func_94533_d() {
-    this.drag_state = 0;
-    this.dragged.clear();
-  }
-
   public static boolean canDrag(final Slot slot, final ItemStack is, final boolean always_true) {
     boolean can = slot == null || !slot.getHasStack();
     if (slot != null && slot.getHasStack() && is != null && is.isItemEqual(slot.getStack())
@@ -216,60 +203,7 @@
   public ItemStack slotClick(final int i, final int param, final int type, final EntityPlayer ep) {
     ItemStack ret = null;
     final InventoryPlayer player_inv = ep.inventory;
-    if (type == 5) {
-      final int prev_drag_state = this.drag_state;
-      this.drag_state = func_94532_c(param);
-      if ((prev_drag_state != 1 || this.drag_state != 2) && prev_drag_state != this.drag_state)
-        func_94533_d();
-      else if (player_inv.getItemStack() == null)
-        func_94533_d();
-      else if (this.drag_state == 0) {
-        this.field_94535_f = func_94529_b(param);
-        if (func_94528_d(this.field_94535_f)) {
-          this.drag_state = 1;
-          this.dragged.clear();
-        } else
-          func_94533_d();
-      } else if (this.drag_state == 1) {
-        final Slot slot = (Slot) this.inventorySlots.get(i);
-        if (slot != null && canDrag(slot, player_inv.getItemStack(), true)
-            && slot.isItemValid(player_inv.getItemStack())
-            && player_inv.getItemStack().stackSize > this.dragged.size() && func_94531_b(slot))
-          this.dragged.add(slot);
-      } else if (this.drag_state == 2) {
-        if (!this.dragged.isEmpty()) {
-          ItemStack player_stack = player_inv.getItemStack().copy();
-          int i1 = player_inv.getItemStack().stackSize;
-          final Iterator<Slot> iterator = this.dragged.iterator();
-          while (iterator.hasNext()) {
-            final Slot slot = iterator.next();
-            if (slot != null && canDrag(slot, player_inv.getItemStack(), true)
-                && slot.isItemValid(player_inv.getItemStack())
-                && player_inv.getItemStack().stackSize >= this.dragged.size()
-                && func_94531_b(slot)) {
-              final ItemStack slot_stack = player_stack.copy();
-              final int j1 = slot.getHasStack() ? slot.getStack().stackSize : 0;
-              func_94525_a(this.dragged, this.field_94535_f, slot_stack, j1);
-              if (!(slot instanceof SlotUnlimited)
-                  && slot_stack.stackSize > slot_stack.getMaxStackSize())
-                slot_stack.stackSize = slot_stack.getMaxStackSize();
-              if (slot_stack.stackSize > slot.getSlotStackLimit())
-                slot_stack.stackSize = slot.getSlotStackLimit();
-              i1 -= slot_stack.stackSize - j1;
-              slot.putStack(slot_stack);
-            }
-          }
-          player_stack.stackSize = i1;
-          if (player_stack.stackSize <= 0)
-            player_stack = null;
-          player_inv.setItemStack(player_stack);
-        }
-        func_94533_d();
-      } else
-        func_94533_d();
-    } else if (this.drag_state != 0)
-      func_94533_d();
-    else if ((type == 0 || type == 1) && (param == 0 || param == 1)) {
+    if ((type == 0 || type == 1) && (param == 0 || param == 1)) {
       if (i == -999) {
         if (player_inv.getItemStack() != null && i == -999) {
           if (param == 0) {
@@ -404,31 +338,6 @@
         slot.onPickupFromSlot(ep, drop_stack);
         ep.dropPlayerItemWithRandomChoice(drop_stack, true);
       }
-    } else if (type == 6 && i >= 0) {
-      final Slot slot = (Slot) this.inventorySlots.get(i);
-      final ItemStack player_stack = player_inv.getItemStack();
-      if (player_stack != null && (slot == null || !slot.getHasStack() || !slot.canTakeStack(ep))) {
-        final int i1 = param == 0 ? 0 : this.inventorySlots.size() - 1;
-        final int l1 = param == 0 ? 1 : -1;
-        for (int i2 = 0; i2 < 2; ++i2)
-          for (int j2 = i1; j2 >= 0 && j2 < this.inventorySlots.size()
-              && player_stack.stackSize < player_stack.getMaxStackSize(); j2 += l1) {
-            final Slot slot3 = (Slot) this.inventorySlots.get(j2);
-            if (slot3.getHasStack() && canDrag(slot3, player_stack, true) && slot3.canTakeStack(ep)
-                && func_94530_a(player_stack, slot3)
-                && (i2 != 0 || slot3.getStack().stackSize != slot3.getStack().getMaxStackSize())) {
-              final int k1 =
-                  Math.min(player_stack.getMaxStackSize() - player_stack.stackSize,
-                      slot3.getStack().stackSize);
-              final ItemStack itemstack2 = slot3.decrStackSize(k1);
-              player_stack.stackSize += k1;
-              if (itemstack2.stackSize <= 0)
-                slot3.putStack(null);
-              slot3.onPickupFromSlot(ep, itemstack2);
-            }
-          }
-      }
-      detectAndSendChanges();
     }
     return ret;
   }
--- a/com/yogpc/mc_lib/GuiWorkbench.java
+++ b/com/yogpc/mc_lib/GuiWorkbench.java
@@ -24,7 +24,7 @@
   @Override
   protected void drawGuiContainerForegroundLayer(final int p_146979_1_, final int p_146979_2_) {
     this.fontRenderer.drawString(
-        this.tile.isInvNameLocalized() ? this.tile.getInvName() : StatCollector
+        StatCollector
             .translateToLocal(this.tile.getInvName()), 8, 6, 0x404040);
     this.fontRenderer.drawString(StatCollector.translateToLocal("container.inventory"), 8,
         this.ySize - 96 + 2, 0x404040);
@@ -34,7 +34,7 @@
   protected void drawGuiContainerBackgroundLayer(final float p_146976_1_, final int p_146976_2_,
       final int p_146976_3_) {
     GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
-    this.mc.renderEngine.bindTexture(gui);
+    this.mc.renderEngine.bindTexture(this.mc.renderEngine.getTexture(gui));
     final int xf = this.width - this.xSize >> 1;
     final int yf = this.height - this.ySize >> 1;
     drawTexturedModalRect(xf, yf, 0, 0, this.xSize, this.ySize);
--- a/com/yogpc/mc_lib/InvUtils.java
+++ b/com/yogpc/mc_lib/InvUtils.java
@@ -37,9 +37,12 @@
         iii = ii;
     } else
       iii = ii;
-    if (iii instanceof ISidedInventory)
-      a = ((ISidedInventory) iii).getAccessibleSlotsFromSide(fd.ordinal());
-    else {
+    if (iii instanceof ISidedInventory) {
+      a = new int[((ISidedInventory) iii).getSizeInventorySide(fd)];
+      final int from = ((ISidedInventory) iii).getStartInventorySide(fd);
+      for (int i = 0; i < a.length; i++)
+        a[i] = from + i;
+    } else {
       a = new int[iii.getSizeInventory()];
       for (int i = 0; i < a.length; i++)
         a[i] = i;
@@ -47,11 +50,6 @@
     int buf, rem = is1.stackSize;
     final List<Integer> e = new ArrayList<Integer>();
     for (final int i : a) {
-      if (iii instanceof ISidedInventory) {
-        if (!((ISidedInventory) iii).canInsertItem(i, is1, fd.ordinal()))
-          continue;
-      } else if (!iii.isStackValidForSlot(i, is1))
-        continue;
       final ItemStack is2 = iii.getStackInSlot(i);
       if (is2 == null) {
         e.add(new Integer(i));
--- a/com/yogpc/mc_lib/ProxyClient.java
+++ b/com/yogpc/mc_lib/ProxyClient.java
@@ -64,16 +64,16 @@
     if (currentScreen == null || currentScreen.allowUserInput)
       for (final Key k : Key.values())
         if (k.binding instanceof KeyBinding) {
-          if (GameSettings.isKeyDown((KeyBinding) k.binding))
+          if (((KeyBinding) k.binding).pressed)
             this.key |= 1 << k.ordinal();
         } else
           switch (k) {
             case forward:
-              if (GameSettings.isKeyDown(Minecraft.getMinecraft().gameSettings.keyBindForward))
+              if (Minecraft.getMinecraft().gameSettings.keyBindForward.pressed)
                 this.key |= 1 << k.ordinal();
               break;
             case jump:
-              if (GameSettings.isKeyDown(Minecraft.getMinecraft().gameSettings.keyBindJump))
+              if (Minecraft.getMinecraft().gameSettings.keyBindJump.pressed)
                 this.key |= 1 << k.ordinal();
               break;
             default:
--- a/com/yogpc/mc_lib/TileWorkbench.java
+++ b/com/yogpc/mc_lib/TileWorkbench.java
@@ -169,21 +169,11 @@
   }
 
   @Override
-  public boolean isStackValidForSlot(final int p_94041_1_, final ItemStack p_94041_2_) {
-    return true;
-  }
-
-  @Override
   public String getInvName() {
     return "tile.WorkbenchPlus.name";
   }
 
   @Override
-  public boolean isInvNameLocalized() {
-    return false;
-  }
-
-  @Override
   public int getInventoryStackLimit() {
     return Integer.MAX_VALUE;
   }
--- a/com/yogpc/nei/WBPRecipeHandler.java
+++ b/com/yogpc/nei/WBPRecipeHandler.java
@@ -75,7 +75,7 @@
   @Override
   public void drawBackground(final GuiContainerManager gui, final int recipe) {
     GL11.glColor4f(1, 1, 1, 1);
-    gui.bindTexture(getGuiTexture());
+    gui.bindTextureByName(getGuiTexture());
     gui.drawTexturedModalRect(0, 0, 5, 11, 166, 121);
   }
 
--- a/com/yogpc/qp/BlockBreaker.java
+++ b/com/yogpc/qp/BlockBreaker.java
@@ -49,6 +49,10 @@
     setHardness(3.5F);
     setStepSound(soundStoneFootstep);
     setBlockName("BreakerPlus");
+    this.blockIndexInTexture = 96;
+    this.furnaceTopIcon = 97;
+    this.horizontal = 100;
+    this.vectrial = 101;
   }
 
   @Override
@@ -74,11 +78,8 @@
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister par1IconRegister) {
-    this.blockIndexInTexture = par1IconRegister.registerIcon("yogpstop_qp:plusstone_side");
-    this.furnaceTopIcon = par1IconRegister.registerIcon("yogpstop_qp:plusstone_top");
-    this.horizontal = par1IconRegister.registerIcon("yogpstop_qp:breaker_front_horizontal");
-    this.vectrial = par1IconRegister.registerIcon("yogpstop_qp:breaker_front_vertical");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
@@ -108,7 +109,7 @@
     final Block b = Block.blocksList[w.getBlockId(tx, ty, tz)];
     if (b == null || b.isAirBlock(w, x, y, z))
       return;
-    final EntityPlayer player = CommonProxy.getPlayer(w);
+    final EntityPlayer player = CommonProxy.getPlayer(w, x, y, z);
     b.onBlockHarvested(w, tx, ty, tz, meta, player);
     if (b.removeBlockByPlayer(w, player, tx, ty, tz))
       b.onBlockDestroyedByPlayer(w, tx, ty, tz, meta);
@@ -139,10 +140,10 @@
 
   @Override
   public void onBlockPlacedBy(final World world, final int x, final int y, final int z, final EntityLiving ent) {
-    EnchantmentHelper.init((IEnchantableTile) world.getBlockTileEntity(x, y, z),
-        is.getEnchantmentTagList());
+    EnchantmentHelper.init((IEnchantableTile) world.getBlockTileEntity(x, y, z), ent.getHeldItem()
+        .getEnchantmentTagList());
     world.setBlockMetadataWithNotify(x, y, z,
-        BlockPistonBase.determineOrientation(world, x, y, z, ent), 2);
+        BlockPistonBase.determineOrientation(world, x, y, z, (EntityPlayer) ent));
   }
 
   static void setDispenserDefaultDirection(final World w, final int x, final int y, final int z) {
--- a/com/yogpc/qp/BlockFrame.java
+++ b/com/yogpc/qp/BlockFrame.java
@@ -23,12 +23,13 @@
     setHardness(0.5F);
     setTickRandomly(true);
     setBlockName("qpFrame");
+    this.blockIndexInTexture = 35;
   }
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister ir) {
-    this.blockIndexInTexture = ir.registerIcon("yogpstop_qp:blockFrame");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
--- a/com/yogpc/qp/BlockInfMJSrc.java
+++ b/com/yogpc/qp/BlockInfMJSrc.java
@@ -29,12 +29,7 @@
     setStepSound(soundStoneFootstep);
     setCreativeTab(QuarryPlus.ct);
     setBlockName("InfMJSrc");
-  }
-
-  @Override
-  @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister ir) {
-    this.blockIndexInTexture = ir.registerIcon("portal");
+    this.blockIndexInTexture = Block.portal.blockIndexInTexture;
   }
 
   @Override
--- a/com/yogpc/qp/BlockLaser.java
+++ b/com/yogpc/qp/BlockLaser.java
@@ -29,13 +29,16 @@
 public class BlockLaser extends BlockContainer {
 
   @SideOnly(Side.CLIENT)
-  private int textureTop, textureBottom;
+  private final int textureTop, textureBottom;
 
   public BlockLaser(final int id) {
     super(id, Material.iron);
     setHardness(10F);
     setCreativeTab(QuarryPlus.ct);
     setBlockName("LaserPlus");
+    this.blockIndexInTexture = 112;
+    this.textureBottom = 113;
+    this.textureTop = 114;
   }
 
   @Override
@@ -80,16 +83,14 @@
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister par1IconRegister) {
-    this.textureTop = par1IconRegister.registerIcon("yogpstop_qp:laser_top");
-    this.textureBottom = par1IconRegister.registerIcon("yogpstop_qp:laser_bottom");
-    this.blockIndexInTexture = par1IconRegister.registerIcon("yogpstop_qp:laser_side");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
   public void onBlockPlacedBy(final World world, final int x, final int y, final int z, final EntityLiving ent) {
     EnchantmentHelper.init((IEnchantableTile) world.getBlockTileEntity(x, y, z),
-        is.getEnchantmentTagList());
+        ent.getHeldItem().getEnchantmentTagList());
   }
 
   private final ArrayList<ItemStack> drop = new ArrayList<ItemStack>();
--- a/com/yogpc/qp/BlockMarker.java
+++ b/com/yogpc/qp/BlockMarker.java
@@ -26,12 +26,17 @@
 import cpw.mods.fml.relauncher.SideOnly;
 
 public class BlockMarker extends BlockContainer {
+  private static final ForgeDirection[] m2fd = new ForgeDirection[] {ForgeDirection.DOWN,
+      ForgeDirection.EAST, ForgeDirection.WEST, ForgeDirection.SOUTH, ForgeDirection.NORTH,
+      ForgeDirection.UP};
+  private static final byte[] s2m = new byte[] {0, 5, 4, 3, 2, 1};
 
   public BlockMarker(final int id) {
     super(id, Material.circuits);
     setLightValue(0.5F);
     setCreativeTab(QuarryPlus.ct);
     setBlockName("MarkerPlus");
+    this.blockIndexInTexture = 16;
   }
 
   @Override
@@ -44,7 +49,7 @@
       final int z) {
     final float w = 0.15F;
     final float h = 0.65F;
-    final ForgeDirection dir = ForgeDirection.getOrientation(world.getBlockMetadata(x, y, z));
+    final ForgeDirection dir = m2fd[world.getBlockMetadata(x, y, z)];
     switch (dir) {
       case DOWN:
         setBlockBounds(0.5F - w, 1F - h, 0.5F - w, 0.5F + w, 1F, 0.5F + w);
@@ -86,14 +91,14 @@
   @Override
   public boolean canPlaceBlockOnSide(final World world, final int x, final int y, final int z,
       final int side) {
-    final ForgeDirection dir = ForgeDirection.getOrientation(side);
+    final ForgeDirection dir =  m2fd[side];
     return world.isBlockSolidOnSide(x - dir.offsetX, y - dir.offsetY, z - dir.offsetZ, dir.getOpposite());
   }
 
   @Override
   public int onBlockPlaced(final World world, final int x, final int y, final int z,
       final int side, final float par6, final float par7, final float par8, final int meta) {
-    return side;
+    return s2m[side];
   }
 
   @Override
@@ -104,7 +109,7 @@
 
   private void dropTorchIfCantStay(final World w, final int x, final int y, final int z) {
     final int m = w.getBlockMetadata(x, y, z);
-    if (!canPlaceBlockOnSide(w, x, y, z, m)) {
+    if (!canPlaceBlockOnSide(w, x, y, z, m2fd[m].ordinal())) {
       dropBlockAsItem(w, x, y, z, m, 0);
       w.setBlock(x, y, z, 0);
     }
@@ -153,7 +158,7 @@
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister par1IconRegister) {
-    this.blockIndexInTexture = par1IconRegister.registerIcon("yogpstop_qp:marker");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 }
--- a/com/yogpc/qp/BlockMiningWell.java
+++ b/com/yogpc/qp/BlockMiningWell.java
@@ -37,6 +37,11 @@
     setCreativeTab(QuarryPlus.ct);
     setStepSound(soundStoneFootstep);
     setBlockName("MiningWellPlus");
+    this.blockIndexInTexture = 48;
+    this.textureFront = 49;
+    this.textureBack = 50;
+    this.textureTop = 51;
+    this.texW = 52;
   }
 
   @Override
@@ -46,12 +51,8 @@
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister par1IconRegister) {
-    this.textureFront = par1IconRegister.registerIcon("yogpstop_qp:miningwell_front");
-    this.blockIndexInTexture = par1IconRegister.registerIcon("yogpstop_qp:miningwell");
-    this.textureBack = par1IconRegister.registerIcon("yogpstop_qp:miningwell_back");
-    this.textureTop = par1IconRegister.registerIcon("yogpstop_qp:miningwell_top");
-    this.texW = par1IconRegister.registerIcon("yogpstop_qp:miningwell_top_w");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
@@ -84,8 +85,8 @@
   public void onBlockPlacedBy(final World w, final int x, final int y, final int z, final EntityLiving el) {
     super.onBlockPlacedBy(w, x, y, z, el);
     final ForgeDirection orientation = get2dOrientation(el.posX, el.posZ, x, z);
-    w.setBlockMetadataWithNotify(x, y, z, orientation.getOpposite().ordinal(), 1);
-    EnchantmentHelper.init((IEnchantableTile) w.getBlockTileEntity(x, y, z), is.getEnchantmentTagList());
+    w.setBlockMetadataWithNotify(x, y, z, orientation.getOpposite().ordinal());
+    EnchantmentHelper.init((IEnchantableTile) w.getBlockTileEntity(x, y, z), el.getHeldItem().getEnchantmentTagList());
   }
 
   private static ForgeDirection get2dOrientation(final double x1, final double z1, final double x2,
--- a/com/yogpc/qp/BlockMover.java
+++ b/com/yogpc/qp/BlockMover.java
@@ -28,6 +28,9 @@
     setHardness(1.2F);
     setCreativeTab(QuarryPlus.ct);
     setBlockName("EnchantMover");
+    this.blockIndexInTexture = 32;
+    this.textureBottom = 33;
+    this.textureTop = 34;
   }
 
   @Override
@@ -46,10 +49,8 @@
 
   @SideOnly(Side.CLIENT)
   @Override
-  public void registerIcons(final IconRegister par1IconRegister) {
-    this.blockIndexInTexture = par1IconRegister.registerIcon("yogpstop_qp:mover");
-    this.textureTop = par1IconRegister.registerIcon("yogpstop_qp:mover_top");
-    this.textureBottom = par1IconRegister.registerIcon("yogpstop_qp:mover_bottom");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
--- a/com/yogpc/qp/BlockPlacer.java
+++ b/com/yogpc/qp/BlockPlacer.java
@@ -46,6 +46,10 @@
     setHardness(3.5F);
     setStepSound(soundStoneFootstep);
     setBlockName("PlacerPlus");
+    this.blockIndexInTexture = 96;
+    this.furnaceTopIcon = 97;
+    this.horizontal = 98;
+    this.vectrial = 99;
   }
 
   @Override
@@ -71,11 +75,8 @@
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister par1IconRegister) {
-    this.blockIndexInTexture = par1IconRegister.registerIcon("yogpstop_qp:plusstone_side");
-    this.furnaceTopIcon = par1IconRegister.registerIcon("yogpstop_qp:plusstone_top");
-    this.horizontal = par1IconRegister.registerIcon("yogpstop_qp:placer_front_horizontal");
-    this.vectrial = par1IconRegister.registerIcon("yogpstop_qp:placer_front_vertical");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
@@ -127,7 +128,7 @@
       ty = y + fd2.offsetY;
       tz = z + fd2.offsetZ;
 
-      final EntityPlayer player = CommonProxy.getPlayer(world);
+      final EntityPlayer player = CommonProxy.getPlayer(world, x, y, z);
       ItemStack is = null;
       int i = 0;
       for (; i < tile.getSizeInventory(); i++) {
@@ -188,7 +189,7 @@
   @Override
   public void onBlockPlacedBy(final World par1World, final int par2, final int par3, final int par4, final EntityLiving par5EntityLiving) {
     final int l =
-        BlockPistonBase.determineOrientation(par1World, par2, par3, par4, par5EntityLiving);
+        BlockPistonBase.determineOrientation(par1World, par2, par3, par4, (EntityPlayer) par5EntityLiving);
     par1World.setBlockMetadataWithNotify(par2, par3, par4, l);
   }
 
@@ -229,7 +230,6 @@
         }
       }
 
-      world.func_96440_m(x, y, z, b);
     }
 
     super.breakBlock(world, x, y, z, b, meta);
--- a/com/yogpc/qp/BlockPlainPipe.java
+++ b/com/yogpc/qp/BlockPlainPipe.java
@@ -17,12 +17,13 @@
     this.maxY = 1.0;
     this.maxZ = 0.75;
     setBlockName("qpPlainPipe");
+    this.blockIndexInTexture = 36;
   }
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister ir) {
-    this.blockIndexInTexture = ir.registerIcon("yogpstop_qp:blockPlainPipe");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
--- a/com/yogpc/qp/BlockPump.java
+++ b/com/yogpc/qp/BlockPump.java
@@ -28,13 +28,18 @@
 
 public class BlockPump extends BlockContainer {
 
-  private int textureTop, textureBottom, textureSide, texW, texC;
+  private final int textureTop, textureBottom, textureSide, texW, texC;
 
   public BlockPump(final int id) {
     super(id, Material.iron);
     setHardness(5F);
     setCreativeTab(QuarryPlus.ct);
     setBlockName("PumpPlus");
+    this.textureSide = 64;
+    this.textureBottom = 65;
+    this.textureTop = 66;
+    this.texC = 67;
+    this.texW = 68;
   }
 
   @Override
@@ -70,12 +75,8 @@
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister par1IconRegister) {
-    this.textureTop = par1IconRegister.registerIcon("yogpstop_qp:pump_top");
-    this.textureBottom = par1IconRegister.registerIcon("yogpstop_qp:pump_bottom");
-    this.textureSide = par1IconRegister.registerIcon("yogpstop_qp:pump_side");
-    this.texW = par1IconRegister.registerIcon("yogpstop_qp:pump_top_w");
-    this.texC = par1IconRegister.registerIcon("yogpstop_qp:pump_top_c");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   private final ArrayList<ItemStack> drop = new ArrayList<ItemStack>();
@@ -106,7 +107,7 @@
   @Override
   public void onBlockPlacedBy(final World w, final int x, final int y, final int z, final EntityLiving el) {
     super.onBlockPlacedBy(w, x, y, z, el);
-    EnchantmentHelper.init((IEnchantableTile) w.getBlockTileEntity(x, y, z), is.getEnchantmentTagList());
+    EnchantmentHelper.init((IEnchantableTile) w.getBlockTileEntity(x, y, z), el.getHeldItem().getEnchantmentTagList());
   }
 
   @Override
--- a/com/yogpc/qp/BlockQuarry.java
+++ b/com/yogpc/qp/BlockQuarry.java
@@ -37,6 +37,12 @@
     setStepSound(soundStoneFootstep);
     setCreativeTab(QuarryPlus.ct);
     setBlockName("QuarryPlus");
+    this.blockIndexInTexture = 0;
+    this.textureFront = 1;
+    this.textureTop = 2;
+    this.texNNB = 3;
+    this.texMF = 4;
+    this.texBB = 5;
   }
 
   private final ArrayList<ItemStack> drop = new ArrayList<ItemStack>();
@@ -101,13 +107,8 @@
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister par1IconRegister) {
-    this.blockIndexInTexture = par1IconRegister.registerIcon("yogpstop_qp:quarry");
-    this.textureTop = par1IconRegister.registerIcon("yogpstop_qp:quarry_top");
-    this.textureFront = par1IconRegister.registerIcon("yogpstop_qp:quarry_front");
-    this.texBB = par1IconRegister.registerIcon("yogpstop_qp:quarry_top_bb");
-    this.texNNB = par1IconRegister.registerIcon("yogpstop_qp:quarry_top_nnb");
-    this.texMF = par1IconRegister.registerIcon("yogpstop_qp:quarry_top_mf");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
@@ -119,9 +120,9 @@
   public void onBlockPlacedBy(final World w, final int x, final int y, final int z, final EntityLiving el) {
     super.onBlockPlacedBy(w, x, y, z, el);
     final ForgeDirection orientation = get2dOrientation(el.posX, el.posZ, x, z);
-    w.setBlockMetadataWithNotify(x, y, z, orientation.getOpposite().ordinal(), 1);
+    w.setBlockMetadataWithNotify(x, y, z, orientation.getOpposite().ordinal());
     ((TileQuarry) w.getBlockTileEntity(x, y, z)).requestTicket();
-    EnchantmentHelper.init((IEnchantableTile) w.getBlockTileEntity(x, y, z), is.getEnchantmentTagList());
+    EnchantmentHelper.init((IEnchantableTile) w.getBlockTileEntity(x, y, z), el.getHeldItem().getEnchantmentTagList());
   }
 
   private static ForgeDirection get2dOrientation(final double x1, final double z1, final double x2,
--- a/com/yogpc/qp/BlockRefinery.java
+++ b/com/yogpc/qp/BlockRefinery.java
@@ -39,12 +39,13 @@
     setHardness(5F);
     setCreativeTab(QuarryPlus.ct);
     setBlockName("RefineryPlus");
+    this.blockIndexInTexture = 17;
   }
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister ir) {
-    this.blockIndexInTexture = ir.registerIcon("yogpstop_qp:refineryDummy");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
@@ -80,11 +81,11 @@
   @Override
   public void onBlockPlacedBy(final World w, final int x, final int y, final int z, final EntityLiving el) {
     super.onBlockPlacedBy(w, x, y, z, el);
-    EnchantmentHelper.init((IEnchantableTile) w.getBlockTileEntity(x, y, z), is.getEnchantmentTagList());
+    EnchantmentHelper.init((IEnchantableTile) w.getBlockTileEntity(x, y, z), el.getHeldItem().getEnchantmentTagList());
     final ForgeDirection[] da =
         {ForgeDirection.NORTH, ForgeDirection.EAST, ForgeDirection.SOUTH, ForgeDirection.WEST};
     final int di = MathHelper.floor_double((el.rotationYaw + 45.0) / 90.0) & 3;
-    w.setBlockMetadataWithNotify(x, y, z, da[di].ordinal(), 1);
+    w.setBlockMetadataWithNotify(x, y, z, da[di].ordinal());
   }
 
   private static void consumeItem(final EntityPlayer ep, final ItemStack stack) {
@@ -129,17 +130,17 @@
       final int meta = world.getBlockMetadata(x, y, z);
       switch (ForgeDirection.values()[meta]) {
         case WEST:
-          world.setBlockMetadataWithNotify(x, y, z, ForgeDirection.SOUTH.ordinal(), 3);
+          world.setBlockMetadataWithNotify(x, y, z, ForgeDirection.SOUTH.ordinal());
           break;
         case EAST:
-          world.setBlockMetadataWithNotify(x, y, z, ForgeDirection.NORTH.ordinal(), 3);
+          world.setBlockMetadataWithNotify(x, y, z, ForgeDirection.NORTH.ordinal());
           break;
         case NORTH:
-          world.setBlockMetadataWithNotify(x, y, z, ForgeDirection.WEST.ordinal(), 3);
+          world.setBlockMetadataWithNotify(x, y, z, ForgeDirection.WEST.ordinal());
           break;
         case SOUTH:
         default:
-          world.setBlockMetadataWithNotify(x, y, z, ForgeDirection.EAST.ordinal(), 3);
+          world.setBlockMetadataWithNotify(x, y, z, ForgeDirection.EAST.ordinal());
           break;
       }
       return true;
--- a/com/yogpc/qp/CommonProxy.java
+++ b/com/yogpc/qp/CommonProxy.java
@@ -17,5 +17,38 @@
 import net.minecraft.entity.player.EntityPlayer;
 
 public class CommonProxy {
+  private static EntityPlayer player;
+
   public void registerTextures() {}
+
+  private static EntityPlayer createNewPlayer(World w) {
+    EntityPlayer p = new EntityPlayer(w) {
+      @Override
+      public void sendChatToPlayer(String var1) {}
+
+      @Override
+      public boolean canCommandSenderUseCommand(int var1, String var2) {
+        return false;
+      }
+
+      @Override
+      public ChunkCoordinates getPlayerCoordinates() {
+        return null;
+      }
+
+    };
+    p.username = "[QuarryPlus]";
+    return p;
+  }
+
+  public static EntityPlayer getPlayer(World w, int x, int y, int z) {
+    if (player == null)
+      player = createNewPlayer(w);
+    else
+      player.worldObj = w;
+    player.posX = x;
+    player.posY = y;
+    player.posZ = z;
+    return player;
+  }
 }
--- a/com/yogpc/qp/EntityLaser.java
+++ b/com/yogpc/qp/EntityLaser.java
@@ -5,10 +5,10 @@
 import net.minecraft.world.World;
 
 public class EntityLaser extends Entity {
-  public static final int DRILL = 0;
-  public static final int DRILL_HEAD = 1;
-  public static final int BLUE_LASER = 2;
-  public static final int RED_LASER = 3;
+  public static final int DRILL = 21;
+  public static final int DRILL_HEAD = 20;
+  public static final int BLUE_LASER = 19;
+  public static final int RED_LASER = 18;
 
   public double iSize, jSize, kSize;
   public final int texture;
--- a/com/yogpc/qp/ItemTool.java
+++ b/com/yogpc/qp/ItemTool.java
@@ -37,6 +37,9 @@
     setHasSubtypes(true);
     setMaxDamage(0);
     setCreativeTab(QuarryPlus.ct);
+    this.iconIndex = 80;
+    this.ile = 81;
+    this.ils = 82;
   }
 
   @Override
@@ -147,10 +150,8 @@
 
   @Override
   @SideOnly(Side.CLIENT)
-  public void registerIcons(final IconRegister ir) {
-    this.iconIndex = ir.registerIcon("yogpstop_qp:statusChecker");
-    this.ile = ir.registerIcon("yogpstop_qp:listEditor");
-    this.ils = ir.registerIcon("yogpstop_qp:liquidSelector");
+  public String getTextureFile() {
+    return "/mods/yogpstop_qp/textures/textures.png";
   }
 
   @Override
--- a/com/yogpc/qp/PowerManager.java
+++ b/com/yogpc/qp/PowerManager.java
@@ -34,7 +34,7 @@
     if (c.containsKey(name)) {
       Property prop = c.get(name);
       if (prop.getType() == null) {
-        prop = new Property(prop.getName(), prop.getString(), Property.Type.DOUBLE);
+        prop = new Property(prop.getName(), prop.value, Property.Type.DOUBLE);
         c.put(name, prop);
       }
       return prop.getDouble(def);
--- a/com/yogpc/qp/QuarryPlus.java
+++ b/com/yogpc/qp/QuarryPlus.java
@@ -76,23 +76,6 @@
         l.destructor();
   }
 
-  @ForgeSubscribe
-  @SideOnly(Side.CLIENT)
-  public void loadTextures(final TextureStitchEvent.Pre evt) {
-    if (evt.map.textureType == 0) {
-      final TextureMap map = evt.map;
-      RenderEntityLaser.icons = new int[4];
-      RenderEntityLaser.icons[EntityLaser.DRILL] =
-          map.registerIcon("yogpstop_qp:blockDrillTexture");
-      RenderEntityLaser.icons[EntityLaser.DRILL_HEAD] =
-          map.registerIcon("yogpstop_qp:blockDrillHeadTexture");
-      RenderEntityLaser.icons[EntityLaser.RED_LASER] =
-          map.registerIcon("yogpstop_qp:blockRedLaser");
-      RenderEntityLaser.icons[EntityLaser.BLUE_LASER] =
-          map.registerIcon("yogpstop_qp:blockBlueLaser");
-    }
-  }
-
   @Mod.PreInit
   public void preInit(final FMLPreInitializationEvent event) {
     final Configuration cfg = new Configuration(event.getSuggestedConfigurationFile());
@@ -212,6 +195,8 @@
     proxy.registerTextures();
     PacketHandler.registerStaticHandler(TileMarker.class);
   }
+  
+  public static final int WILDCARD_VALUE = 32767;
 
   public static class BlockData {
     public final String name;
--- a/com/yogpc/qp/TileBasic.java
+++ b/com/yogpc/qp/TileBasic.java
@@ -273,11 +273,6 @@
   }
 
   @Override
-  public boolean isInvNameLocalized() {
-    return false;
-  }
-
-  @Override
   public int getInventoryStackLimit() {
     return 0;
   }
@@ -295,9 +290,4 @@
 
   @Override
   public void closeChest() {}
-
-  @Override
-  public boolean isStackValidForSlot(final int p_94041_1_, final ItemStack p_94041_2_) {
-    return false;
-  }
 }
--- a/com/yogpc/qp/TileBreaker.java
+++ b/com/yogpc/qp/TileBreaker.java
@@ -96,11 +96,6 @@
   }
 
   @Override
-  public boolean isInvNameLocalized() {
-    return false;
-  }
-
-  @Override
   public int getInventoryStackLimit() {
     return 0;
   }
@@ -118,9 +113,4 @@
 
   @Override
   public void closeChest() {}
-
-  @Override
-  public boolean isStackValidForSlot(final int p_94041_1_, final ItemStack p_94041_2_) {
-    return false;
-  }
 }
--- a/com/yogpc/qp/TilePlacer.java
+++ b/com/yogpc/qp/TilePlacer.java
@@ -76,11 +76,6 @@
   }
 
   @Override
-  public boolean isInvNameLocalized() {
-    return false;
-  }
-
-  @Override
   public int getInventoryStackLimit() {
     return 64;
   }
@@ -98,11 +93,6 @@
   public void closeChest() {}
 
   @Override
-  public boolean isStackValidForSlot(final int i, final ItemStack itemstack) {
-    return true;
-  }
-
-  @Override
   public void readFromNBT(final NBTTagCompound par1NBTTagCompound) {
     super.readFromNBT(par1NBTTagCompound);
     final NBTTagList nbttaglist = par1NBTTagCompound.getTagList("Items");
--- a/com/yogpc/qp/TilePump.java
+++ b/com/yogpc/qp/TilePump.java
@@ -142,7 +142,7 @@
               this.zCoord + fd.offsetZ);
       if (te instanceof ITankContainer)
         for (final String s : this.mapping[fd.ordinal()]) {
-          pZ = this.liquids.indexOf(LiquidDictionary.getCanonicalLiquid(s));
+          pZ = indexOf(LiquidDictionary.getLiquid(s, 0));
           if (pZ == -1)
             continue;
           fs = this.liquids.get(pZ);
@@ -526,7 +526,7 @@
       for (int i = 0; i < this.liquids.size(); i++) {
         final LiquidStack s = this.liquids.get(i);
         ret[i + 1] =
-            StatCollector.translateToLocalFormatted("yog.pump.liquid", LiquidDictionary.findLiquidName(s)
+            StatCollector.translateToLocalFormatted("yog.pump.liquid", findLiquidName(s)
                 , Integer.toString(s.amount));
       }
     } else
@@ -547,7 +547,7 @@
   public LiquidStack drain(final ForgeDirection fd, final LiquidStack resource, final boolean doDrain) {
     if (resource == null)
       return null;
-    final int index = this.liquids.indexOf(resource);
+    final int index = indexOf(resource);
     if (index == -1)
       return null;
     final LiquidStack fs = this.liquids.get(index);
@@ -585,10 +585,10 @@
       int index;
       LiquidStack fs;
       for (final String s : this.mapping[fd.ordinal()]) {
-        fs = LiquidDictionary.getCanonicalLiquid(s);
+        fs = LiquidDictionary.getLiquid(s, 0);
         if (fs == null)
           continue;
-        index = this.liquids.indexOf(fs);
+        index = indexOf(fs);
         if (index != -1)
           ret.add(new LiquidTank(this.liquids.get(index), Integer.MAX_VALUE));
         else
@@ -623,7 +623,7 @@
       fs = LiquidDictionary.getLiquid(s, maxDrain);
       if (fs == null)
         continue;
-      index = this.liquids.indexOf(fs);
+      index = indexOf(fs);
       if (index == -1)
         continue;
       return drain(fd, this.liquids.get(index), doDrain);
@@ -666,7 +666,7 @@
       else if ((b == Block.lavaStill || b == Block.lavaMoving) && meta == 0)
         fs = new LiquidStack(Block.lavaStill, LiquidContainerRegistry.BUCKET_VOLUME);
       if (fs != null) {
-        final int index = this.liquids.indexOf(fs);
+        final int index = indexOf(fs);
         if (index != -1)
           this.liquids.get(index).amount += fs.amount;
         else
@@ -724,4 +724,16 @@
     final TileBasic tb = G_connected();
     return tb == null ? 0 : tb.powerRequest();
   }
+
+  private static String findLiquidName(final LiquidStack fs) {
+    for (final Map.Entry<String, LiquidStack> entry : LiquidDictionary.getLiquids().entrySet())
+      if (fs.isLiquidEqual(entry.getValue())) return entry.getKey();
+    return null;
+  }
+
+  private int indexOf(final LiquidStack key) {
+    for (int i = 0; i < this.liquids.size(); i++)
+      if (this.liquids.get(i).isLiquidEqual(key)) return i;
+    return -1;
+  }
 }
--- a/com/yogpc/qp/client/ClientProxy.java
+++ b/com/yogpc/qp/client/ClientProxy.java
@@ -37,5 +37,6 @@
     RenderingRegistry.registerBlockHandler(RenderLaserBlock.INSTANCE);
     RenderingRegistry.registerBlockHandler(RenderMarker.INSTANCE);
     RenderingRegistry.registerBlockHandler(RenderFrame.INSTANCE);
+    MinecraftForgeClient.preloadTexture("/mods/yogpstop_qp/textures/textures.png");
   }
 }
--- a/com/yogpc/qp/client/GuiError.java
+++ b/com/yogpc/qp/client/GuiError.java
@@ -20,28 +20,30 @@
 import cpw.mods.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class GuiError extends GuiErrorScreen {
-  private final GuiScreen parent;
+public class GuiError extends GuiScreenA {
+  private final String message1, message2;
 
   public GuiError(final GuiScreen par1GuiScreen, final String par2Str, final String par3Str) {
-    super(par2Str, par3Str);
-    this.parent = par1GuiScreen;
+    super(par1GuiScreen);
+    this.message1 = par2Str;
+    this.message2 = par3Str;
   }
 
   @Override
   protected void actionPerformed(final GuiButton par1GuiButton) {
-    Minecraft.getMinecraft().displayGuiScreen(this.parent);
+    showParent();
   }
 
   @Override
-  public boolean doesGuiPauseGame() {
-    return false;
+  public void initGui() {
+    this.controlList.add(new GuiSmallButton(0, this.width / 2 - 75, this.height / 6 + 96, StatCollector.translateToLocal("gui.done")));
   }
 
   @Override
-  public void updateScreen() {
-    super.updateScreen();
-    if (!this.mc.thePlayer.isEntityAlive() || this.mc.thePlayer.isDead)
-      this.mc.thePlayer.closeScreen();
+  public void drawScreen(final int par1, final int par2, final float par3) {
+    drawDefaultBackground();
+    drawCenteredString(this.fontRenderer, this.message1, this.width / 2, 70, 0xFFFFFF);
+    drawCenteredString(this.fontRenderer, this.message2, this.width / 2, 90, 0xFFFFFF);
+    super.drawScreen(par1, par2, par3);
   }
 }
--- a/com/yogpc/qp/client/GuiInfMJSrc.java
+++ b/com/yogpc/qp/client/GuiInfMJSrc.java
@@ -129,7 +129,7 @@
   @Override
   protected void drawGuiContainerBackgroundLayer(final float f, final int i, final int j) {
     GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
-    this.mc.renderEngine.bindTexture(gui);
+    this.mc.renderEngine.bindTexture(this.mc.renderEngine.getTexture(gui));
     drawTexturedModalRect(this.width - 176 >> 1, this.height - 214 >> 1, 0, 0, this.xSize,
         this.ySize);
     this.eng.drawTextBox();
--- a/com/yogpc/qp/client/GuiMover.java
+++ b/com/yogpc/qp/client/GuiMover.java
@@ -63,7 +63,7 @@
   @Override
   protected void drawGuiContainerBackgroundLayer(final float f, final int i, final int j) {
     GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
-    this.mc.renderEngine.bindTexture(gui);
+    this.mc.renderEngine.bindTexture(this.mc.renderEngine.getTexture(gui));
     drawTexturedModalRect(this.width - this.xSize >> 1, this.height - this.ySize >> 1, 0, 0,
         this.xSize, this.ySize);
   }
--- a/com/yogpc/qp/client/GuiPlacer.java
+++ b/com/yogpc/qp/client/GuiPlacer.java
@@ -40,7 +40,7 @@
   @Override
   protected void drawGuiContainerForegroundLayer(final int par1, final int par2) {
     final String s =
-        this.tile.isInvNameLocalized() ? this.tile.getInvName() : StatCollector
+        StatCollector
             .translateToLocal(this.tile.getInvName());
     this.fontRenderer.drawString(s, this.xSize / 2 - this.fontRenderer.getStringWidth(s) / 2,
         6, 4210752);
@@ -51,7 +51,7 @@
   @Override
   protected void drawGuiContainerBackgroundLayer(final float par1, final int par2, final int par3) {
     GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
-    this.mc.renderEngine.bindTexture(tex);
+    this.mc.renderEngine.bindTexture(this.mc.renderEngine.getTexture(tex));
     final int k = (this.width - this.xSize) / 2;
     final int l = (this.height - this.ySize) / 2;
     drawTexturedModalRect(k, l, 0, 0, this.xSize, this.ySize);
--- a/com/yogpc/qp/client/RenderEntityLaser.java
+++ b/com/yogpc/qp/client/RenderEntityLaser.java
@@ -16,7 +16,6 @@
 
 @SideOnly(Side.CLIENT)
 public class RenderEntityLaser extends Render {
-  public static int[] icons;
   public static RenderEntityLaser INSTANCE = new RenderEntityLaser();
 
   private RenderEntityLaser() {}
@@ -47,10 +46,10 @@
   }
 
   private static void doRender(final RenderEngine tm, final double i, final double j,
-      final double k, final double iSize, final double jSize, final double kSize, final int tex) {
+      final double k, final double iSize, final double jSize, final double kSize, final int texture) {
     GL11.glPushMatrix();
     GL11.glTranslated(i, j, k);
-    tm.bindTexture("/terrain.png");
+    tm.bindTexture(tm.getTexture("/mods/yogpstop_qp/textures/textures.png"));
     final Tessellator t = Tessellator.instance;
     final RenderBlocks rb = new RenderBlocks();
     for (int iBase = 0; iBase < iSize; ++iBase)
@@ -61,9 +60,6 @@
           final double remainZ = kSize - kBase;
           GL11.glPushMatrix();
           GL11.glTranslatef(iBase, jBase, kBase);
-          int texture = icons[tex];
-          if (texture == null)
-            texture = Block.sand.getBlockTextureFromSide(0);
           t.startDrawingQuads();
           rb.setRenderBounds(0, 0, 0, remainX > 1.0 ? 1.0 : remainX, remainY > 1.0 ? 1.0 : remainY,
               remainZ > 1.0 ? 1.0 : remainZ);
--- a/com/yogpc/qp/client/RenderLaser.java
+++ b/com/yogpc/qp/client/RenderLaser.java
@@ -39,7 +39,7 @@
     GL11.glRotatef((float) (360 - (Math.atan2(dz, dx) * 180.0 / Math.PI + 180.0)), 0, 1, 0);
     GL11.glRotatef((float) (-Math.atan2(dy, Math.sqrt(total * total - dy * dy)) * 180.0 / Math.PI),
         0, 0, 1);
-    tm.bindTexture(tex);
+    tm.bindTexture(tm.getTexture(tex));
     int i = 0;
     while (i <= total - 1) {
       box[b].render(1F / 16);
--- a/com/yogpc/qp/client/RenderMarker.java
+++ b/com/yogpc/qp/client/RenderMarker.java
@@ -146,10 +146,10 @@
     x += 0.5D;
     z += 0.5D;
 
-    double minU = i.getInterpolatedU(7);
-    double minV = i.getInterpolatedV(7);
-    double maxU = i.getInterpolatedU(9);
-    double maxV = i.getInterpolatedV(9);
+    double minU = (((i & 0xF) << 4) + 7) / 256F;
+    double minV = ((i & 0xF0) + 7) / 256F;
+    double maxU = (((i & 0xF) << 4) + 9) / 256F;
+    double maxV = ((i & 0xF0) + 9) / 256F;
 
     tessellator.setBrightness(block
         .getMixedBrightnessForBlock(iblockaccess, xCoord, yCoord, zCoord));
@@ -190,10 +190,10 @@
 
     i = block.getBlockTexture(iblockaccess, xCoord, yCoord, zCoord, 0);
 
-    minU = i.getMinU();
-    maxU = i.getMaxU();
-    minV = i.getMinV();
-    maxV = i.getMaxV();
+    minU = ((i & 0xF) << 4) / 256F;
+    minV = (i & 0xF0) / 256F;
+    maxU = (((i & 0xF) << 4) + 0x10) / 256F;
+    maxV = ((i & 0xF0) + 0x10) / 256F;
 
     if (m == 5 || m == 4 || m == 3 || m == 0) {
       tessellator.addVertexWithUV(x + frontX[m][0][0], y + frontX[m][1][0], z + frontX[m][2][0],
--- a/com/yogpc/qp/client/RenderRefinery.java
+++ b/com/yogpc/qp/client/RenderRefinery.java
@@ -177,7 +177,7 @@
           GL11.glPushMatrix();
           GL11.glTranslatef(tx, ty, tz);
         }
-        bindTextureByName(liquid.getTextureSheet());
+        bindTextureByName(Block.blocksList[liquid.itemID].getTextureFile());
         final ItemStack stack = liquid.asItemStack();
         setColor(stack.getItem().getColorFromItemStack(stack, 0));
         GL11.glCallList(list[(int) (liquid.amount / buf * 99)]);
@@ -196,13 +196,10 @@
       return diplayLists;
     diplayLists = new int[100];
     Block baseBlock;
-    int texture;
     if (fluidStack.itemID > -1) {
       baseBlock = Block.blocksList[fluidStack.itemID];
-      texture = fluidStack.getRenderingIcon();
     } else {
       baseBlock = Block.waterStill;
-      texture = fluidStack.getRenderingIcon();
     }
     GL11.glDisable(GL11.GL_LIGHTING);
     GL11.glDisable(GL11.GL_BLEND);
@@ -214,17 +211,17 @@
       tessellator.startDrawingQuads();
       renderBlocks.setRenderBounds(0.01, 0, 0.01, 0.99, (float) s / 100, 0.99);
       renderBlocks.renderBottomFace(baseBlock, 0, 0, 0,
-          texture != null ? texture : baseBlock.getBlockTextureFromSide(0));
+          baseBlock.getBlockTextureFromSide(0));
       renderBlocks.renderTopFace(baseBlock, 0, 0, 0,
-          texture != null ? texture : baseBlock.getBlockTextureFromSide(1));
+          baseBlock.getBlockTextureFromSide(1));
       renderBlocks.renderEastFace(baseBlock, 0, 0, 0,
-          texture != null ? texture : baseBlock.getBlockTextureFromSide(2));
+          baseBlock.getBlockTextureFromSide(2));
       renderBlocks.renderWestFace(baseBlock, 0, 0, 0,
-          texture != null ? texture : baseBlock.getBlockTextureFromSide(3));
+          baseBlock.getBlockTextureFromSide(3));
       renderBlocks.renderNorthFace(baseBlock, 0, 0, 0,
-          texture != null ? texture : baseBlock.getBlockTextureFromSide(4));
+          baseBlock.getBlockTextureFromSide(4));
       renderBlocks.renderSouthFace(baseBlock, 0, 0, 0,
-          texture != null ? texture : baseBlock.getBlockTextureFromSide(5));
+          baseBlock.getBlockTextureFromSide(5));
       tessellator.draw();
       GL11.glEndList();
     }
