--- a/com/yogpc/ip/ItemArmorElectric.java
+++ b/com/yogpc/ip/ItemArmorElectric.java
@@ -14,6 +14,7 @@
 import cpw.mods.fml.common.registry.GameRegistry;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
+import net.minecraft.client.renderer.texture.IconRegister;
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.player.EntityPlayer;
@@ -33,12 +34,17 @@
 		super(id, EnumHelper.addArmorMaterial("ELECTRIC", 33, new int[] { 3, 8, 6, 3 }, 0), YogpstopLib.proxy.addNewArmourRendererPrefix("electric"), 1);
 		this.setUnlocalizedName("electric_armor");
 		this.setCreativeTab(CreativeTabs.tabCombat);
-		this.setTextureName("yogpstop_qp:electric_armor");
 		GameRegistry.registerItem(this, "qpArmor");
 		this.setMaxDamage(100);
 		this.setMaxStackSize(1);
 	}
 
+	@Override
+	@SideOnly(Side.CLIENT)
+	public void registerIcons(IconRegister ir) {
+		itemIcon = ir.registerIcon("yogpstop_qp:electric_armor");
+	}
+
 	private static final double maxFallSpeedOnHover = 0.35;
 
 	private void useJetpack(EntityPlayer p, ItemStack jetpack, boolean hover) {
--- a/com/yogpc/mc_lib/YogpstopLib.java
+++ b/com/yogpc/mc_lib/YogpstopLib.java
@@ -15,7 +15,7 @@
 	@SidedProxy(clientSide = "com.yogpc.mc_lib.ProxyClient", serverSide = "com.yogpc.mc_lib.ProxyCommon")
 	public static ProxyCommon proxy;
 
-	@Mod.EventHandler
+	@Mod.Init
 	public void init(FMLInitializationEvent event) {
 		new ItemArmorElectric(21478);// TODO IC2Plus
 	}
--- a/com/yogpc/qp/APowerTile.java
+++ b/com/yogpc/qp/APowerTile.java
@@ -3,13 +3,13 @@
 import com.yogpc.mc_lib.APacketTile;
 
 import net.minecraft.nbt.NBTTagCompound;
-import net.minecraft.world.World;
 import net.minecraftforge.common.ForgeDirection;
 import buildcraft.api.power.IPowerReceptor;
 import buildcraft.api.power.IPowerProvider;
+import buildcraft.api.power.PowerFramework;
 
 public abstract class APowerTile extends APacketTile implements IPowerReceptor {
-	private final IPowerProvider pp = new IPowerProvider(this, Type.MACHINE);
+	private final IPowerProvider pp = PowerFramework.currentFramework.createPowerProvider();
 	private double stored, recv_max, stored_max;
 
 	@Override
@@ -31,8 +31,13 @@
 	public final void doWork() {}
 
 	@Override
-	public final World getWorld() {
-		return this.worldObj;
+	public final int powerRequest(ForgeDirection d) {
+		return (int) Math.ceil(Math.min(this.recv_max, this.stored_max - this.stored));
+	}
+
+	@Override
+	public final void setPowerProvider(IPowerProvider p) {
+		return;
 	}
 
 	@Override
@@ -41,7 +46,7 @@
 		this.stored = nbttc.getDouble("storedEnergy");
 		this.stored_max = nbttc.getDouble("MAX_stored");
 		this.recv_max = nbttc.getDouble("MAX_receive");
-		this.pp.configure(0, (float) this.recv_max, 0, (float) this.stored_max);
+		this.pp.configure(0, 0, (int) this.recv_max, 0, (int) this.stored_max);
 	}
 
 	@Override
@@ -77,6 +82,6 @@
 	public final void configure(double max, double maxstored) {
 		this.recv_max = max;
 		this.stored_max = maxstored;
-		this.pp.configure(0, (float) this.recv_max, 0, (float) this.stored_max);
+		this.pp.configure(0, 0, (int) this.recv_max, 0, (int) this.stored_max);
 	}
 }
--- a/com/yogpc/qp/BlockFrame.java
+++ b/com/yogpc/qp/BlockFrame.java
@@ -7,6 +7,7 @@
 import cpw.mods.fml.relauncher.SideOnly;
 import net.minecraft.block.Block;
 import net.minecraft.block.material.Material;
+import net.minecraft.client.renderer.texture.IconRegister;
 import net.minecraft.entity.Entity;
 import net.minecraft.item.Item;
 import net.minecraft.util.AxisAlignedBB;
@@ -23,7 +24,12 @@
 		setSides(true, true, true, true, true, true);
 		setHardness(0.5F);
 		setTickRandomly(true);
-		this.setTextureName("buildcraft:blockFrame");// TODO buildcraft resource
+	}
+
+	@Override
+	@SideOnly(Side.CLIENT)
+	public void registerIcons(IconRegister ir) {
+		this.blockIcon = ir.registerIcon("buildcraft:blockFrame");// TODO buildcraft resource
 	}
 
 	@Override
--- a/com/yogpc/qp/BlockInfMJSrc.java
+++ b/com/yogpc/qp/BlockInfMJSrc.java
@@ -19,9 +19,12 @@
 
 import net.minecraft.block.BlockContainer;
 import net.minecraft.block.material.Material;
+import net.minecraft.client.renderer.texture.IconRegister;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.world.World;
+import cpw.mods.fml.relauncher.SideOnly;
+import cpw.mods.fml.relauncher.Side;
 
 public class BlockInfMJSrc extends BlockContainer {
 
@@ -32,7 +35,12 @@
 		setStepSound(soundStoneFootstep);
 		setCreativeTab(QuarryPlus.ct);
 		setUnlocalizedName("InfMJSrc");
-		setTextureName("portal");
+	}
+
+	@Override
+	@SideOnly(Side.CLIENT)
+	public void registerIcons(IconRegister ir) {
+		this.blockIcon = ir.registerIcon("portal");
 	}
 
 	@Override
--- a/com/yogpc/qp/BlockPlainPipe.java
+++ b/com/yogpc/qp/BlockPlainPipe.java
@@ -4,7 +4,10 @@
 
 import net.minecraft.block.Block;
 import net.minecraft.block.material.Material;
+import net.minecraft.client.renderer.texture.IconRegister;
 import net.minecraft.item.Item;
+import cpw.mods.fml.relauncher.SideOnly;
+import cpw.mods.fml.relauncher.Side;
 
 public class BlockPlainPipe extends Block {
 	public BlockPlainPipe(int id) {
@@ -15,7 +18,12 @@
 		this.maxX = 0.75;
 		this.maxY = 1.0;
 		this.maxZ = 0.75;
-		this.setTextureName("buildcraft:blockPlainPipe");// TODO buildcraft resource
+	}
+
+	@Override
+	@SideOnly(Side.CLIENT)
+	public void registerIcons(IconRegister ir) {
+		this.blockIcon = ir.registerIcon("buildcraft:blockPlainPipe");// TODO buildcraft resource
 	}
 
 	@Override
--- a/com/yogpc/qp/BlockRefinery.java
+++ b/com/yogpc/qp/BlockRefinery.java
@@ -21,6 +21,7 @@
 
 import net.minecraft.block.BlockContainer;
 import net.minecraft.block.material.Material;
+import net.minecraft.client.renderer.texture.IconRegister;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.Item;
@@ -32,6 +33,8 @@
 import net.minecraftforge.liquids.LiquidContainerRegistry;
 import net.minecraftforge.liquids.LiquidStack;
 import net.minecraftforge.liquids.ITankContainer;
+import cpw.mods.fml.relauncher.SideOnly;
+import cpw.mods.fml.relauncher.Side;
 
 public class BlockRefinery extends BlockContainer {
 
@@ -40,7 +43,12 @@
 		setHardness(5F);
 		setCreativeTab(QuarryPlus.ct);
 		setUnlocalizedName("RefineryPlus");
-		setTextureName("buildcraft:refineryBack");// TODO buildcraft resource
+	}
+
+	@Override
+	@SideOnly(Side.CLIENT)
+	public void registerIcons(IconRegister ir) {
+		this.blockIcon = ir.registerIcon("buildcraft:refineryBack");// TODO buildcraft resource
 	}
 
 	@Override
@@ -94,7 +102,7 @@
 
 	private static boolean fill(ITankContainer tank, ForgeDirection side, EntityPlayer player) {
 		ItemStack current = player.getCurrentEquippedItem();
-		LiquidStack liquid = LiquidContainerRegistry.getFluidForFilledItem(current);
+		LiquidStack liquid = LiquidContainerRegistry.getLiquidForFilledItem(current);
 		if (liquid != null) {
 			int used = tank.fill(side, liquid, true);
 			if (used > 0) {
--- a/com/yogpc/qp/QuarryPlus.java
+++ b/com/yogpc/qp/QuarryPlus.java
@@ -99,7 +99,7 @@
 	@ForgeSubscribe
 	@SideOnly(Side.CLIENT)
 	public void loadTextures(TextureStitchEvent.Pre evt) {
-		if (evt.map.getTextureType() == 0) {
+		if (evt.map.textureType == 0) {
 			TextureMap map = evt.map;
 			RenderEntityLaser.icons = new Icon[4];
 			RenderEntityLaser.icons[EntityLaser.DRILL] = map.registerIcon("buildcraft:blockDrillTexture");// TODO buildcraft resource
@@ -109,7 +109,7 @@
 		}
 	}
 
-	@Mod.EventHandler
+	@Mod.PreInit
 	public void preInit(FMLPreInitializationEvent event) {
 		Configuration cfg = new Configuration(event.getSuggestedConfigurationFile());
 		int[] id = new int[13];
@@ -162,10 +162,10 @@
 	}
 
 	private static final void addAssemblyRecipe(int a, ItemStack b, ItemStack... c) {
-		AssemblyRecipe.assemblyRecipes.add(new AssemblyRecipe(a, b, c));
+		AssemblyRecipe.assemblyRecipes.add(new AssemblyRecipe(c, a, b));
 	}
 
-	@Mod.EventHandler
+	@Mod.Init
 	public void init(FMLInitializationEvent event) {
 		GameRegistry.registerItem(itemTool, "qpTool");
 		GameRegistry.registerBlock(blockQuarry, ItemBlockQuarry.class, "QuarryPlus");
--- a/com/yogpc/qp/TileBasic.java
+++ b/com/yogpc/qp/TileBasic.java
@@ -44,8 +44,8 @@
 import net.minecraft.tileentity.TileEntityChest;
 import net.minecraft.world.World;
 import net.minecraftforge.common.ForgeDirection;
-import buildcraft.api.power.IPowerEmitter;
-import buildcraft.api.transport.IPipeTile;
+import buildcraft.api.transport.IPipeConnection;
+import buildcraft.api.transport.IPipeEntry;
 
 public abstract class TileBasic extends APowerTile implements IInventory, IEnchantableTile {
 	protected ForgeDirection pump = ForgeDirection.UNKNOWN;
@@ -205,21 +205,22 @@
 		return moved;
 	}
 
+	private static final Class engcls = ReflectionHelper.getClass("buildcraft.energy.TileEngine");
+
 	static int injectToNearTile(World w, int x, int y, int z, ItemStack is) {
-		List<IPipeTile> pp = new LinkedList<IPipeTile>();
+		List<TileEntity> pp = new LinkedList<TileEntity>();
 		List<ForgeDirection> ppd = new LinkedList<ForgeDirection>();
 		List<IInventory> pi = new LinkedList<IInventory>();
 		List<ForgeDirection> pid = new LinkedList<ForgeDirection>();
 		for (ForgeDirection d : ForgeDirection.VALID_DIRECTIONS) {
 			TileEntity t = w.getBlockTileEntity(x + d.offsetX, y + d.offsetY, z + d.offsetZ);
-			if ((t instanceof IInventory) && !(t instanceof IPowerEmitter) && addToIInv((IInventory) t, is, d.getOpposite(), false) > 0) {
+			if ((t instanceof IInventory) && (engcls == null || !engcls.isInstance(t)) && addToIInv((IInventory) t, is, d.getOpposite(), false) > 0) {
 				pi.add((IInventory) t);
 				pid.add(d.getOpposite());
 			}
-			if (t instanceof IPipeTile) {
-				IPipeTile p = (IPipeTile) t;
-				if (p.getPipeType() != IPipeTile.PipeType.ITEM || !p.isPipeConnected(d.getOpposite())) continue;
-				pp.add(p);
+			if (t instanceof IPipeEntry && ((IPipeEntry) t).acceptItems()) {
+				if (t instanceof IPipeConnection && !((IPipeConnection) t).isPipeConnected(d.getOpposite())) continue;
+				pp.add(t);
 				ppd.add(d.getOpposite());
 			}
 		}
@@ -229,7 +230,8 @@
 		}
 		if (pp.size() > 0) {
 			int i = w.rand.nextInt(pp.size());
-			return pp.get(i).injectItem(is, true, ppd.get(i));
+			((IPipeEntry) pp.get(i)).entityEntering(is, ppd.get(i));
+			return is.stackSize;
 		}
 		return 0;
 	}
--- a/com/yogpc/qp/TileInfMJSrc.java
+++ b/com/yogpc/qp/TileInfMJSrc.java
@@ -21,6 +21,7 @@
 import java.io.DataOutputStream;
 import java.lang.reflect.Method;
 
+import buildcraft.api.power.IPowerProvider;
 import buildcraft.api.power.IPowerReceptor;
 
 import com.google.common.io.ByteArrayDataInput;
@@ -65,8 +66,8 @@
 				}
 			} catch (Exception e) {}
 			if (te instanceof IPowerReceptor) {
-				IPowerProvider pr = ((IPowerReceptor) te).getPowerProvider(d.getOpposite());
-				if (pr != null) pr.receiveEnergy(Type.ENGINE, this.power, d.getOpposite());
+				IPowerProvider pr = ((IPowerReceptor) te).getPowerProvider();
+				if (pr != null) pr.receiveEnergy(this.power, d.getOpposite());
 			}
 		}
 		this.cInterval = this.interval;
--- a/com/yogpc/qp/TileLaser.java
+++ b/com/yogpc/qp/TileLaser.java
@@ -179,9 +179,9 @@
 		}
 	}
 
-	public static final String[] LASER_TEXTURES = new String[] { new String("buildcraft", "textures/entities/laser_1.png"),// TODO buildcraft resource
-			new String("buildcraft", "textures/entities/laser_2.png"), new String("buildcraft", "textures/entities/laser_3.png"),// TODO buildcraft resource
-			new String("buildcraft", "textures/entities/laser_4.png"), new String("buildcraft", "textures/entities/stripes.png") };// TODO buildcraft resource
+	public static final String[] LASER_TEXTURES = new String[] { "/gfx/buildcraft/entities/laser_1.png",// TODO buildcraft resource
+			"/gfx/buildcraft/entities/laser_2.png", "/gfx/buildcraft/entities/laser_3.png",// TODO buildcraft resource
+			"/gfx/buildcraft/entities/laser_4.png", "/gfx/buildcraft/entities/stripes.png" };// TODO buildcraft resource
 
 	public String getTexture() {
 		double avg = this.pa / 100;
--- a/com/yogpc/qp/TilePump.java
+++ b/com/yogpc/qp/TilePump.java
@@ -38,7 +38,7 @@
 import net.minecraft.world.World;
 import net.minecraft.world.chunk.storage.ExtendedBlockStorage;
 import net.minecraftforge.common.ForgeDirection;
-import net.minecraftforge.liquids.Fluid;
+import net.minecraftforge.liquids.ILiquidTank;
 import net.minecraftforge.liquids.LiquidContainerRegistry;
 import net.minecraftforge.liquids.LiquidDictionary;
 import net.minecraftforge.liquids.LiquidStack;
@@ -134,7 +134,7 @@
 			te = this.worldObj.getBlockTileEntity(this.xCoord + fd.offsetX, this.yCoord + fd.offsetY, this.zCoord + fd.offsetZ);
 			if (te instanceof ITankContainer) {
 				for (String s : this.mapping[fd.ordinal()]) {
-					pZ = this.liquids.indexOf(LiquidDictionary.getLiquidStack(s, 0));
+					pZ = this.liquids.indexOf(LiquidDictionary.getCanonicalLiquid(s));
 					if (pZ == -1) continue;
 					fs = this.liquids.get(pZ);
 					fs.amount -= ((ITankContainer) te).fill(fd.getOpposite(), fs, true);
@@ -485,7 +485,7 @@
 		if (this.liquids.size() > 0) {
 			ret[0] = StatCollector.translateToLocal("chat.pumpcontain");
 			for (int i = 0; i < this.liquids.size(); i++) {
-				ret[i + 1] = new StringBuilder().append("    ").append(this.liquids.get(i).getFluid().getLocalizedName()).append(": ")
+				ret[i + 1] = new StringBuilder().append("    ").append(LiquidDictionary.findLiquidName(this.liquids.get(i))).append(": ")
 						.append(this.liquids.get(i).amount).append("mB").toString();
 			}
 		} else {
@@ -519,6 +519,10 @@
 	}
 
 	@Override
+	public LiquidStack drain(int id, int maxDrain, boolean doDrain) {
+		return null;// TODO
+	}
+
 	public LiquidStack drain(ForgeDirection fd, LiquidStack resource, boolean doDrain) {
 		if (resource == null) return null;
 		int index = this.liquids.indexOf(resource);
@@ -534,13 +538,8 @@
 	}
 
 	@Override
-	public boolean canFill(ForgeDirection from, Fluid fluid) {
-		return false;
-	}
-
-	@Override
-	public boolean canDrain(ForgeDirection from, Fluid fluid) {
-		return true;
+	public int fill(int id, LiquidStack resource, boolean doFill) {
+		return 0;
 	}
 
 	@Override
@@ -549,8 +548,8 @@
 		LinkedList<LiquidTank> ret = new LinkedList<LiquidTank>();
 		if (this.mapping[fd.ordinal()].size() <= 0) {
 			if (this.liquids.size() <= 0) {
-				for (Integer i : LiquidDictionary.getLiquids().values())
-					ret.add(new LiquidTank(new LiquidStack(i, 0), Integer.MAX_VALUE));
+				for (LiquidStack i : LiquidDictionary.getLiquids().values())
+					ret.add(new LiquidTank(i, Integer.MAX_VALUE));
 			} else {
 				for (LiquidStack fs : this.liquids)
 					ret.add(new LiquidTank(fs, Integer.MAX_VALUE));
@@ -559,7 +558,7 @@
 			int index;
 			LiquidStack fs;
 			for (String s : this.mapping[fd.ordinal()]) {
-				fs = LiquidDictionary.getLiquidStack(s, 0);
+				fs = LiquidDictionary.getCanonicalLiquid(s);
 				if (fs == null) continue;
 				index = this.liquids.indexOf(fs);
 				if (index != -1) ret.add(new LiquidTank(this.liquids.get(index), Integer.MAX_VALUE));
@@ -570,6 +569,16 @@
 	}
 
 	@Override
+	public ILiquidTank getTank(ForgeDirection d, LiquidStack t) {
+		ILiquidTank[] ilda = getTanks(d);
+		if (ilda == null) return null;
+		if (t == null) return ilda[0];
+		for (ILiquidTank ild : ilda)
+			if (ild.getLiquid().isLiquidEqual(t)) return ild;
+		return null;
+	}
+
+	@Override
 	public LiquidStack drain(ForgeDirection fd, int maxDrain, boolean doDrain) {
 		if (fd.ordinal() < 0 || fd.ordinal() >= this.mapping.length) return drain(ForgeDirection.UP, maxDrain, doDrain);
 		if (this.mapping[fd.ordinal()].size() <= 0) {
@@ -579,7 +588,7 @@
 		int index;
 		LiquidStack fs;
 		for (String s : this.mapping[fd.ordinal()]) {
-			fs = LiquidDictionary.getLiquidStack(s, maxDrain);
+			fs = LiquidDictionary.getLiquid(s, maxDrain);
 			if (fs == null) continue;
 			index = this.liquids.indexOf(fs);
 			if (index == -1) continue;
@@ -603,7 +612,7 @@
 	 * 
 	 * */
 	static final boolean isLiquid(Block b, boolean s, World w, int x, int y, int z, int m) {
-		if (b instanceof ILiquid) return !s || ((ILiquid) b).canDrain(w, x, y, z);
+		if (b instanceof ILiquid) return !s || ((ILiquid) b).stillLiquidMeta() == m;
 		if (b == Block.waterStill || b == Block.waterMoving || b == Block.lavaStill || b == Block.lavaMoving) return !s || m == 0;
 		return false;
 	}
@@ -613,12 +622,12 @@
 		int meta = this.ebses[bx >> 4][bz >> 4][this.py >> 4].getExtBlockMetadata(bx & 0xF, this.py & 0xF, bz & 0xF);
 		if (isLiquid(b, false, null, 0, 0, 0, 0)) {
 			LiquidStack fs = null;
-			if (b instanceof ILiquid && ((ILiquid) b).canDrain(this.worldObj, bx + this.xOffset, this.py, bz + this.zOffset)) {
-				fs = ((ILiquid) b).drain(this.worldObj, bx + this.xOffset, this.py, bz + this.zOffset, true);
+			if (b instanceof ILiquid && ((ILiquid) b).stillLiquidMeta() == meta) {
+				fs = new LiquidStack(((ILiquid) b).stillLiquidId(), LiquidContainerRegistry.BUCKET_VOLUME, ((ILiquid) b).stillLiquidMeta());
 			} else if ((b == Block.waterStill || b == Block.waterMoving) && meta == 0) {
-				fs = new LiquidStack(LiquidDictionary.WATER, LiquidContainerRegistry.BUCKET_VOLUME);
+				fs = new LiquidStack(Block.waterStill, LiquidContainerRegistry.BUCKET_VOLUME);
 			} else if ((b == Block.lavaStill || b == Block.lavaMoving) && meta == 0) {
-				fs = new LiquidStack(LiquidDictionary.LAVA, LiquidContainerRegistry.BUCKET_VOLUME);
+				fs = new LiquidStack(Block.lavaStill, LiquidContainerRegistry.BUCKET_VOLUME);
 			}
 			if (fs != null) {
 				int index = this.liquids.indexOf(fs);
@@ -667,7 +676,13 @@
 	public void doWork() {}
 
 	@Override
-	public World getWorld() {
-		return this.worldObj;
+	public void setPowerProvider(IPowerProvider provider) {
+		return;
+	}
+
+	@Override
+	public int powerRequest(ForgeDirection d) {
+		TileBasic tb = G_connected();
+		return tb == null ? 0 : tb.powerRequest(d);
 	}
 }
--- a/com/yogpc/qp/TileRefinery.java
+++ b/com/yogpc/qp/TileRefinery.java
@@ -23,7 +23,7 @@
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraftforge.common.ForgeDirection;
-import net.minecraftforge.liquids.Fluid;
+import net.minecraftforge.liquids.ILiquidTank;
 import net.minecraftforge.liquids.LiquidContainerRegistry;
 import net.minecraftforge.liquids.LiquidStack;
 import net.minecraftforge.liquids.LiquidTank;
@@ -232,7 +232,6 @@
 		return 0;
 	}
 
-	@Override
 	public LiquidStack drain(ForgeDirection from, LiquidStack resource, boolean doDrain) {
 		// if (resource == null) return null;
 		if (resource.equals(this.res)) return drain(from, resource.amount, doDrain);
@@ -270,13 +269,13 @@
 	}
 
 	@Override
-	public boolean canFill(ForgeDirection from, Fluid fluid) {
-		return true;
+	public int fill(int i, LiquidStack f, boolean d) {
+		return fill(ForgeDirection.UNKNOWN, f, d);
 	}
 
 	@Override
-	public boolean canDrain(ForgeDirection from, Fluid fluid) {
-		return true;
+	public LiquidStack drain(int i, int x, boolean d) {
+		return drain(ForgeDirection.UNKNOWN, x, d);
 	}
 
 	@Override
@@ -285,6 +284,15 @@
 	}
 
 	@Override
+	public ILiquidTank getTank(ForgeDirection direction, LiquidStack type) {
+		if (type == null) return null;
+		if (type.isLiquidEqual(this.src1)) return new LiquidTank(this.src1, this.buf);
+		if (type.isLiquidEqual(this.src2)) return new LiquidTank(this.src2, this.buf);
+		if (type.isLiquidEqual(this.res)) return new LiquidTank(this.res, this.buf);
+		return null;
+	}
+
+	@Override
 	public byte getEfficiency() {
 		return this.efficiency;
 	}
--- a/com/yogpc/qp/client/GuiInfMJSrc.java
+++ b/com/yogpc/qp/client/GuiInfMJSrc.java
@@ -34,7 +34,7 @@
 
 @SideOnly(Side.CLIENT)
 public class GuiInfMJSrc extends GuiScreenA {
-	private static final String gui = new String("yogpstop_qp", "textures/gui/infmjsrc.png");
+	private static final String gui = "/mods/yogpstop_qp/textures/gui/infmjsrc.png";
 	private TileInfMJSrc tile;
 	private GuiTextField eng;
 	private GuiTextField itv;
--- a/com/yogpc/qp/client/GuiMover.java
+++ b/com/yogpc/qp/client/GuiMover.java
@@ -35,7 +35,7 @@
 @SideOnly(Side.CLIENT)
 public class GuiMover extends GuiContainer {
 	public GuiButton b32, b33, b34, b35;
-	private static final String gui = new String("yogpstop_qp", "textures/gui/mover.png");
+	private static final String gui = "/mods/yogpstop_qp/textures/gui/mover.png";
 
 	public GuiMover(EntityPlayer player, World world, int x, int y, int z) {
 		super(null);
--- a/com/yogpc/qp/client/GuiP_List.java
+++ b/com/yogpc/qp/client/GuiP_List.java
@@ -25,7 +25,6 @@
 import net.minecraft.client.gui.GuiButton;
 import net.minecraft.util.StatCollector;
 import net.minecraftforge.common.ForgeDirection;
-import net.minecraftforge.liquids.LiquidDictionary;
 
 @SideOnly(Side.CLIENT)
 public class GuiP_List extends GuiScreenA {
@@ -80,7 +79,6 @@
 			break;
 		case PacketHandler.CtS_REMOVE_MAPPING:
 			String name = this.tile.mapping[this.dir].get(this.oreslot.currentore);
-			if (LiquidDictionary.isFluidRegistered(name)) name = LiquidDictionary.getFluid(name).getLocalizedName();
 			this.mc.displayGuiScreen(new GuiYesNo(this, StatCollector.translateToLocal("tof.deletefluidsure"), name, par1.id));
 			break;
 		default:
--- a/com/yogpc/qp/client/GuiP_Manual.java
+++ b/com/yogpc/qp/client/GuiP_Manual.java
@@ -26,7 +26,6 @@
 import net.minecraft.client.gui.GuiScreen;
 import net.minecraft.client.gui.GuiTextField;
 import net.minecraft.util.StatCollector;
-import net.minecraftforge.liquids.LiquidDictionary;
 
 @SideOnly(Side.CLIENT)
 public class GuiP_Manual extends GuiScreenA {
@@ -55,11 +54,9 @@
 			String name = this.blockid.getText();
 			if (name.length() == 0) return;
 			if (this.tile.mapping[this.targetid].contains(name)) {
-				if (LiquidDictionary.isFluidRegistered(name)) name = LiquidDictionary.getFluid(name).getLocalizedName();
 				this.mc.displayGuiScreen(new GuiError(this, StatCollector.translateToLocal("tof.alreadyerror"), name));
 				return;
 			}
-			if (LiquidDictionary.isFluidRegistered(name)) name = LiquidDictionary.getFluid(name).getLocalizedName();
 			this.mc.displayGuiScreen(new GuiYesNo(this, StatCollector.translateToLocal("tof.addfluidsure"), name, -1));
 			break;
 		case -2:
--- a/com/yogpc/qp/client/GuiP_SlotBlockList.java
+++ b/com/yogpc/qp/client/GuiP_SlotBlockList.java
@@ -77,7 +77,7 @@
 
 	@Override
 	protected void drawSlot(int var1, int var2, int var3, int var4, Tessellator var5) {
-		String name = LiquidDictionary.getFluid(this.blocklist.get(var1)).getLocalizedName();
+		String name = this.blocklist.get(var1);
 		Minecraft.getMinecraft().fontRenderer.drawStringWithShadow(name, (this.parent.width - Minecraft.getMinecraft().fontRenderer.getStringWidth(name)) / 2,
 				var3 + 1, 0xFFFFFF);
 	}
--- a/com/yogpc/qp/client/GuiP_SlotList.java
+++ b/com/yogpc/qp/client/GuiP_SlotList.java
@@ -25,7 +25,6 @@
 import net.minecraft.client.gui.GuiScreen;
 import net.minecraft.client.gui.GuiSlot;
 import net.minecraft.client.renderer.Tessellator;
-import net.minecraftforge.liquids.LiquidDictionary;
 
 @SideOnly(Side.CLIENT)
 public class GuiP_SlotList extends GuiSlot {
@@ -62,7 +61,6 @@
 	@Override
 	protected void drawSlot(int var1, int var2, int var3, int var4, Tessellator var5) {
 		String name = this.target.get(var1);
-		if (LiquidDictionary.isFluidRegistered(name)) name = LiquidDictionary.getFluid(name).getLocalizedName();
 		Minecraft.getMinecraft().fontRenderer.drawStringWithShadow(name,
 				(this.parent.width * 3 / 5 - Minecraft.getMinecraft().fontRenderer.getStringWidth(name)) / 2, var3 + 1, 0xFFFFFF);
 	}
--- a/com/yogpc/qp/client/RenderEntityLaser.java
+++ b/com/yogpc/qp/client/RenderEntityLaser.java
@@ -23,11 +23,6 @@
 	private RenderEntityLaser() {}
 
 	@Override
-	protected String getEntityTexture(Entity entity) {
-		return TextureMap.locationBlocksTexture;
-	}
-
-	@Override
 	public void doRender(Entity e, double i, double j, double k, float f, float f1) {
 		if (e.isDead) return;
 		GL11.glPushMatrix();
@@ -50,7 +45,7 @@
 	private static void doRender(RenderEngine tm, double i, double j, double k, double iSize, double jSize, double kSize, int tex) {
 		GL11.glPushMatrix();
 		GL11.glTranslated(i, j, k);
-		tm.bindTexture(TextureMap.locationBlocksTexture);
+		tm.bindTexture("/terrain.png");
 		Tessellator t = Tessellator.instance;
 		RenderBlocks rb = new RenderBlocks();
 		for (int iBase = 0; iBase < iSize; ++iBase) {
--- a/com/yogpc/qp/client/RenderRefinery.java
+++ b/com/yogpc/qp/client/RenderRefinery.java
@@ -32,7 +32,7 @@
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.Icon;
 import net.minecraft.world.IBlockAccess;
-import net.minecraftforge.liquids.Fluid;
+import net.minecraft.item.ItemStack;
 import net.minecraftforge.liquids.LiquidStack;
 
 import org.lwjgl.opengl.GL11;
@@ -46,7 +46,7 @@
 
 @SideOnly(Side.CLIENT)
 public class RenderRefinery extends TileEntitySpecialRenderer implements ISimpleBlockRenderingHandler {
-	private static final String TEXTURE = new String("yogpstop_qp", "textures/blocks/refinery.png");
+	private static final String TEXTURE = "/mods/yogpstop_qp/textures/blocks/refinery.png";
 	private static final float pixel = (float) (1.0 / 16.0);
 	private static final RenderBlocks renderBlocks = new RenderBlocks();
 	public static final RenderRefinery INSTANCE = new RenderRefinery();
@@ -117,7 +117,7 @@
 
 		GL11.glRotatef(angle, 0, 1, 0);
 
-		bindTexture(TEXTURE);
+		bindTextureByName(TEXTURE);
 
 		GL11.glPushMatrix();
 		GL11.glTranslatef(-0.5F, -0.5F, -0.5F);
@@ -183,31 +183,30 @@
 					GL11.glPushMatrix();
 					GL11.glTranslatef(tx, ty, tz);
 				}
-				bindTexture(TextureMap.locationBlocksTexture);
-				setColor(liquid.getFluid().getColor(liquid));
+				bindTextureByName(liquid.getTextureSheet());
+				ItemStack stack = liquid.asItemStack();
+				setColor(stack.getItem().getColorFromItemStack(stack, 0));
 				GL11.glCallList(list[(int) (liquid.amount / buf * 99)]);
 				if (tx != 0 || ty != 0 || tz != 0) GL11.glPopMatrix();
 			}
 		}
 	}
 
-	private static final Map<Fluid, int[]> stillRenderCache = new HashMap<Fluid, int[]>();
+	private static final Map<LiquidStack, int[]> stillRenderCache = new HashMap<LiquidStack, int[]>();
 
 	private static int[] getLiquidDisplayLists(LiquidStack fluidStack) {
-		Fluid fluid = fluidStack.getFluid();
-		if (fluid == null) return null;
-		Map<Fluid, int[]> cache = stillRenderCache;
-		int[] diplayLists = cache.get(fluid);
+		Map<LiquidStack, int[]> cache = stillRenderCache;
+		int[] diplayLists = cache.get(fluidStack);
 		if (diplayLists != null) return diplayLists;
 		diplayLists = new int[100];
 		Block baseBlock;
 		Icon texture;
-		if (fluid.getBlockID() > -1) {
-			baseBlock = Block.blocksList[fluid.getBlockID()];
-			texture = fluid.getStillIcon();
+		if (fluidStack.itemID > -1) {
+			baseBlock = Block.blocksList[fluidStack.itemID];
+			texture = fluidStack.getRenderingIcon();
 		} else {
 			baseBlock = Block.waterStill;
-			texture = fluid.getStillIcon();
+			texture = fluidStack.getRenderingIcon();
 		}
 		GL11.glDisable(GL11.GL_LIGHTING);
 		GL11.glDisable(GL11.GL_BLEND);
@@ -231,7 +230,7 @@
 		GL11.glEnable(GL11.GL_CULL_FACE);
 		GL11.glEnable(GL11.GL_BLEND);
 		GL11.glEnable(GL11.GL_LIGHTING);
-		cache.put(fluid, diplayLists);
+		cache.put(fluidStack, diplayLists);
 		return diplayLists;
 	}
 
